diff --git a/src/main/scala/common/tile.scala b/src/main/scala/common/tile.scala
index 05821c52..dd1c9032 100644
--- a/src/main/scala/common/tile.scala
+++ b/src/main/scala/common/tile.scala
@@ -6,7 +6,7 @@
 package boom.common
 
 import chisel3._
-import chisel3.util.{RRArbiter, Queue}
+import chisel3.util._
 
 import scala.collection.mutable.{ListBuffer}
 
@@ -243,6 +243,9 @@ class BoomTileModuleImp(outer: BoomTile) extends BaseTileModuleImp(outer){
   lsu.io.hellacache <> hellaCacheArb.io.mem
   outer.dcache.module.io.lsu <> lsu.io.dmem
 
+  // External performance monitor
+  outer.pmSourceNode.bundle <> core.io.pm
+
   // Generate a descriptive string
   val frontendStr = outer.frontend.module.toString
   val coreStr = core.toString
diff --git a/src/main/scala/exu/core.scala b/src/main/scala/exu/core.scala
index 0f8d39e3..c6ab56f8 100644
--- a/src/main/scala/exu/core.scala
+++ b/src/main/scala/exu/core.scala
@@ -37,7 +37,7 @@ import org.chipsalliance.cde.config.Parameters
 import freechips.rocketchip.rocket.Instructions._
 import freechips.rocketchip.tile.{TraceBundle}
 import freechips.rocketchip.rocket.{Causes, PRV, TracedInstruction}
-import freechips.rocketchip.util.{Str, UIntIsOneOf, CoreMonitorBundle}
+import freechips.rocketchip.util.{Str, UIntIsOneOf, CoreMonitorBundle, PmSignals}
 import freechips.rocketchip.devices.tilelink.{PLICConsts, CLINTConsts}
 
 import boom.common._
@@ -61,6 +61,7 @@ class BoomCore()(implicit p: Parameters) extends BoomModule
     val ptw_tlb = new freechips.rocketchip.rocket.TLBPTWIO()
     val trace = Output(new TraceBundle)
     val fcsr_rm = UInt(freechips.rocketchip.tile.FPConstants.RM_SZ.W)
+    val pm = Output(new PmSignals)
   })
 
   io.ptw_tlb := DontCare
@@ -1492,4 +1493,203 @@ class BoomCore()(implicit p: Parameters) extends BoomModule
   } else {
     io.ifu.debug_ftq_idx := DontCare
   }
+
+  // jxy: performance monitor
+  val maskWidth = 4
+  val pidWidth = 12
+  val evtWidth = 2
+  val stageWidth = 2
+  val cntWidth = 4
+  val pipLatency = 14.U
+  val level = Wire(UInt(2.W)) // global
+  val pid   = RegInit(0.U(pidWidth.W))
+  val pc    = Wire(Vec(coreWidth, UInt(vaddrBits.W))) // per instruction
+  val stage = Wire(Vec(coreWidth, UInt(stageWidth.W)))
+  val evt   = RegInit(VecInit(Seq.fill(coreWidth)(0.U(evtWidth.W))))
+  val evm   = Wire(Vec(coreWidth, Vec(maskWidth, Bool())))
+  for (w <- 0 until coreWidth) {
+    io.pm.tag(w)   := RegNext(Cat(pid, level, evt(w)))
+    io.pm.info(w)  := RegNext(Cat(stage(w), pc(w)))
+    io.pm.event(w) := RegNext(evm(w).asUInt)
+  }
+  // Additional info in ROB
+  val next_pc_reg = RegInit(0.U(vaddrBits.W))
+  val last_pc_reg = RegInit(0.U(vaddrBits.W))
+  val last_cnt_reg = RegInit(0.U(cntWidth.W))
+  val last_evm_reg = RegInit(VecInit(Seq.fill(maskWidth)(false.B)))
+  val next_pc = Wire(UInt(vaddrBits.W))
+  val last_pc = Wire(UInt(vaddrBits.W))
+  val last_cnt = Wire(UInt(cntWidth.W))
+  val last_evm = Wire(Vec(maskWidth, Bool()))
+  val next_evm = Wire(Vec(maskWidth, Bool()))
+  val next_stage = Wire(UInt(stageWidth.W))
+  next_pc_reg := next_pc
+  last_pc_reg := last_pc
+  last_cnt_reg := last_cnt
+  last_evm_reg := last_evm
+  next_pc := next_pc_reg
+  last_pc := last_pc_reg
+  last_cnt := Mux(last_cnt_reg =/= 0.U && rob.io.ready, last_cnt_reg - 1.U, last_cnt_reg)
+  last_evm := last_evm_reg
+  next_stage := 0.U
+  next_evm := last_evm
+  for (w <- 0 until coreWidth) {
+    def GetRowIdx(rob_idx: UInt): UInt = {
+      if (coreWidth == 1) return rob_idx
+      else return rob_idx >> log2Ceil(coreWidth).U
+    }
+    def GetBankIdx(rob_idx: UInt): UInt = {
+      if (coreWidth == 1) return 0.U
+      else return rob_idx(log2Ceil(coreWidth)-1, 0).asUInt
+    }
+    def MatchBank(bank_idx: UInt): Bool = (bank_idx === w.U)
+    val rob_stage = RegInit(VecInit(Seq.fill(numRobRows)(0.U(stageWidth.W))))
+    val rob_pid = RegInit(VecInit(Seq.fill(numRobRows)(0.U(pidWidth.W))))
+    val rob_misp = RegInit(VecInit(Seq.fill(numRobRows)(false.B)))
+    val rob_taken = RegInit(VecInit(Seq.fill(numRobRows)(false.B)))
+    val rob_npc = RegInit(VecInit(Seq.fill(numRobRows)(0.U(vaddrBits.W))))
+    val rob_dtm = RegInit(VecInit(Seq.fill(numRobRows)(false.B)))
+    val rob_dcm = RegInit(VecInit(Seq.fill(numRobRows)(false.B)))
+    val rob_cnt = RegInit(VecInit(Seq.fill(numRobRows)(0.U(cntWidth.W))))
+    val com_pc = Wire(UInt(vaddrBits.W))
+    val com_cnt = Wire(UInt(cntWidth.W))
+    val com_evm = Wire(Vec(maskWidth, Bool()))
+    val com_evt = Wire(UInt(evtWidth.W))
+    // Count for redirect scope
+    for (i <- 0 until numRobRows) {
+      when (rob_cnt(i) =/= 0.U && rob.io.ready) {
+        rob_cnt(i) := rob_cnt(i) - 1.U
+      }
+    }
+    // Dispatching
+    for (i <- 0 until coreWidth) {
+      when(dis_valids(i) && MatchBank(GetBankIdx(dis_uops(i).rob_idx))) {
+        rob_stage(GetRowIdx(dis_uops(i).rob_idx)) := 1.U
+        rob_cnt(GetRowIdx(dis_uops(i).rob_idx)) := 0.U
+        rob_misp(GetRowIdx(dis_uops(i).rob_idx)) := false.B
+        rob_dtm(GetRowIdx(dis_uops(i).rob_idx)) := false.B
+        rob_dcm(GetRowIdx(dis_uops(i).rob_idx)) := false.B
+      }
+    }
+    // Issuing
+    for (i <- 0 until exe_units.length) {
+      when (iss_valids(i) && MatchBank(GetBankIdx(iss_uops(i).rob_idx))) {
+        rob_stage(GetRowIdx(iss_uops(i).rob_idx)) := 2.U
+      }
+    }
+    // Record PID from HINT argument
+    for (i <- 0 until rob.numWakeupPorts) {
+      val wb_resp = rob.io.wb_resps(i)
+      val wb_uop = wb_resp.bits.uop
+      when (wb_resp.valid && MatchBank(GetBankIdx(wb_uop.rob_idx))) {
+        rob_stage(GetRowIdx(wb_uop.rob_idx)) := 3.U
+        rob_pid(GetRowIdx(wb_uop.rob_idx)) := wb_resp.bits.data.asUInt
+      }
+    }
+    // Record branches and PCs
+    for (b <- brinfos) {
+      when (b.valid && MatchBank(GetBankIdx(b.uop.rob_idx))) {
+        rob_cnt(GetRowIdx(b.uop.rob_idx)) := pipLatency
+        rob_misp(GetRowIdx(b.uop.rob_idx)) := b.mispredict
+        rob_taken(GetRowIdx(b.uop.rob_idx)) := b.taken
+        rob_npc(GetRowIdx(b.uop.rob_idx)) := b.jalr_target
+      }
+    }
+    // Record DTLB and data cache misses
+    for (i <- 0 until memWidth) {
+      when (io.lsu.tperf(i).tlb_miss.valid && MatchBank(GetBankIdx(io.lsu.tperf(i).tlb_miss.bits))) {
+        rob_dtm(GetRowIdx(io.lsu.tperf(i).tlb_miss.bits)) := true.B
+      }
+      when (io.lsu.tperf(i).cache_miss.valid && MatchBank(GetBankIdx(io.lsu.tperf(i).cache_miss.bits))) {
+        rob_dcm(GetRowIdx(io.lsu.tperf(i).cache_miss.bits)) := true.B
+      }
+    }
+    // Record ITLB and instruction cache miss
+    val itm = Wire(Bool())
+    val icm = Wire(Bool())
+    itm := io.ifu.tperf.tlb_miss(rob.io.commit.uops(w).ftq_idx)
+    icm := io.ifu.tperf.cache_miss(rob.io.commit.uops(w).ftq_idx)
+    for (i <- 0 until w) {
+      when (rob.io.commit.uops(w).ftq_idx === rob.io.commit.uops(i).ftq_idx) {
+        itm := false.B
+        icm := false.B
+      }
+    }
+    // Committing
+    val uop = RegNext(rob.io.commit.uops(w))
+    when (RegNext(rob.io.commit.valids(w))) {
+      rob_stage(GetRowIdx(rob.io.commit.uops(w).rob_idx)) := 0.U
+      rob_misp(GetRowIdx(rob.io.commit.uops(w).rob_idx)) := false.B
+      rob_dtm(GetRowIdx(rob.io.commit.uops(w).rob_idx)) := false.B
+      rob_dcm(GetRowIdx(rob.io.commit.uops(w).rob_idx)) := false.B
+      last_evm := com_evm
+      last_pc := com_pc
+      last_cnt := com_cnt
+      when (uop.is_br && rob_taken(GetRowIdx(uop.rob_idx))) {
+        next_pc := (com_pc.asSInt + ImmGen(uop.imm_packed, IS_B)).asUInt
+      }.elsewhen (uop.is_jal) {
+        next_pc := (com_pc.asSInt + ImmGen(uop.imm_packed, IS_J)).asUInt
+      }.elsewhen (uop.is_jalr) {
+        next_pc := rob_npc(GetRowIdx(uop.rob_idx))
+      }.otherwise {
+        next_pc := com_pc + Mux(uop.is_rvc, 2.U, 4.U)
+      }
+    }
+    when (RegNext(rob.io.flush.valid)) {
+      last_cnt := pipLatency
+      next_pc := io.ifu.redirect_pc
+    }
+    when (MatchBank(GetBankIdx(rob.io.rob_head_idx))) {
+      next_stage := Mux(rob.io.empty, 0.U, rob_stage(GetRowIdx(rob.io.rob_head_idx)))
+      next_evm := Mux(rob.io.empty, VecInit(Seq.fill(maskWidth)(false.B)), com_evm)
+    }
+    // Get stage, event mask and PC
+    io.ifu.debug_ftq_idx(w) := rob.io.commit.uops(w).ftq_idx
+    com_pc := (AlignPCToBoundary(io.ifu.debug_fetch_pc(w), icBlockBytes)
+      + uop.pc_lob - Mux(uop.edge_inst, 2.U, 0.U))(vaddrBits-1,0)
+    com_cnt := rob_cnt(GetRowIdx(uop.rob_idx))
+    stage(w) := 0.U
+    com_evm(0) := 1.B
+    com_evm(1) := RegNext(itm) || RegNext(com_evt) === 1.U && rob_dtm(RegNext(GetRowIdx(rob.io.commit.uops(w).rob_idx)))
+    com_evm(2) := RegNext(icm)
+    com_evm(3) :=
+      RegNext(com_evt) === 1.U && rob_dcm(RegNext(GetRowIdx(rob.io.commit.uops(w).rob_idx))) ||
+      RegNext(com_evt) === 2.U && rob_misp(RegNext(GetRowIdx(rob.io.commit.uops(w).rob_idx)))
+    when (evt(w) === 0.U) {
+      when (last_cnt === 0.U) {
+        pc(w) := next_pc
+        evm(w) := next_evm
+        stage(w) := next_stage
+      }.otherwise {
+        pc(w) := last_pc
+        evm(w) := last_evm
+      }
+    }.otherwise {
+      pc(w) := com_pc
+      evm(w) := com_evm
+    }
+    // Get PID
+    when (
+      rob.io.commit.valids(w) &&
+      rob.io.commit.uops(w).uopc === uopSLLI &&
+      rob.io.commit.uops(w).pdst === 0.U
+    ) {
+      pid := RegNext(rob_pid(GetRowIdx(rob.io.commit.uops(w).rob_idx)))
+    }
+    // Get event tag
+    when (rob.io.commit.uops(w).uopc === uopLD || rob.io.commit.uops(w).uopc === uopSTA) {
+      com_evt := 1.U // EVT::MEM
+    }.elsewhen (rob.io.commit.uops(w).is_br || rob.io.commit.uops(w).is_jal || rob.io.commit.uops(w).is_jalr) {
+      com_evt := 2.U // EVT::BRANCH
+    }.otherwise {
+      com_evt := 3.U // EVT::OTHERS
+    }
+    when (!rob.io.commit.valids(w)) {
+      evt(w) := 0.U // EVT::NULL
+    }.otherwise {
+      evt(w) := com_evt
+    }
+  }
+  // Get privilege level
+  level := csr.io.status.prv
 }
diff --git a/src/main/scala/ifu/frontend.scala b/src/main/scala/ifu/frontend.scala
index 17bf3b2e..cebaaf2b 100644
--- a/src/main/scala/ifu/frontend.scala
+++ b/src/main/scala/ifu/frontend.scala
@@ -286,6 +286,11 @@ class BoomFrontendIO(implicit p: Parameters) extends BoomBundle
   val flush_icache = Output(Bool())
 
   val perf = Input(new FrontendPerfEvents)
+
+  val tperf = Input(new Bundle {
+    val tlb_miss = Vec(ftqSz, Bool())
+    val cache_miss = Vec(ftqSz, Bool())
+  })
 }
 
 /**
@@ -985,6 +990,46 @@ class BoomFrontendModule(outer: BoomFrontend) extends LazyModuleImp(outer)
   ftq.io.debug_ftq_idx := io.cpu.debug_ftq_idx
   io.cpu.debug_fetch_pc := ftq.io.debug_fetch_pc
 
+  // jxy: performance monitor
+  val tlbVadd = RegInit(0.U(vaddrBitsExtended.W))
+  val cachePadd = RegInit(0.U(paddrBits.W))
+  val f3_tlbMiss = withReset(reset.asBool || f3_clear) { Module(new Queue(Bool(), 1, pipe=true, flow=false)) }
+  val f3_cacheMiss = withReset(reset.asBool || f3_clear) { Module(new Queue(Bool(), 1, pipe=true, flow=false)) }
+  val f4_tlbMiss = withReset(reset.asBool || f4_clear) { Module(new Queue(Bool(), 1, pipe=true, flow=false)) }
+  val f4_cacheMiss = withReset(reset.asBool || f4_clear) { Module(new Queue(Bool(), 1, pipe=true, flow=false)) }
+  val ftq_enq_ptr = RegInit(1.U(log2Ceil(ftqSz).W))
+  val ftq_tlbMiss = Reg(Vec(ftqSz, Bool()))
+  val ftq_cacheMiss = Reg(Vec(ftqSz, Bool()))
+  f3_tlbMiss  .io.enq.valid := f3.io.enq.valid
+  f3_cacheMiss.io.enq.valid := f3.io.enq.valid
+  f3_tlbMiss  .io.enq.bits  := (s2_vpc >> pgIdxBits)    ===  tlbVadd
+  f3_cacheMiss.io.enq.bits  := (s2_ppc >> blockOffBits) === (cachePadd >> blockOffBits)
+  f3_tlbMiss  .io.deq.ready := f3.io.deq.ready
+  f3_cacheMiss.io.deq.ready := f3.io.deq.ready
+  f4_tlbMiss  .io.enq.valid := f4.io.enq.valid
+  f4_cacheMiss.io.enq.valid := f4.io.enq.valid
+  f4_tlbMiss  .io.enq.bits  := f3_tlbMiss  .io.deq.bits
+  f4_cacheMiss.io.enq.bits  := f3_cacheMiss.io.deq.bits
+  f4_tlbMiss  .io.deq.ready := f4.io.deq.ready
+  f4_cacheMiss.io.deq.ready := f4.io.deq.ready
+  when (f3_tlbMiss  .io.enq.valid && f3_tlbMiss  .io.enq.bits) { tlbVadd   := 0.U }
+  when (f3_cacheMiss.io.enq.valid && f3_cacheMiss.io.enq.bits) { cachePadd := 0.U }
+  when (tlb   .io.perf.miss)    { tlbVadd   := tlb   .io.perf.vadd }
+  when (icache.io.perf.acquire) { cachePadd := icache.io.perf.padd }
+  when (ftq.io.deq.valid) {
+    ftq_tlbMiss  (ftq.io.deq.bits) := false.B
+    ftq_cacheMiss(ftq.io.deq.bits) := false.B
+  }
+  when (ftq.io.enq.fire) {
+    ftq_tlbMiss  (ftq_enq_ptr) := f4_tlbMiss  .io.deq.bits
+    ftq_cacheMiss(ftq_enq_ptr) := f4_cacheMiss.io.deq.bits
+    ftq_enq_ptr := WrapInc(ftq_enq_ptr, ftqSz)
+  }
+  when (ftq.io.redirect.valid) {
+    ftq_enq_ptr := WrapInc(ftq.io.redirect.bits, ftqSz)
+  }
+  io.cpu.tperf.tlb_miss   := ftq_tlbMiss
+  io.cpu.tperf.cache_miss := ftq_cacheMiss
 
   override def toString: String =
     (BoomCoreStringPrefix("====Overall Frontend Params====") + "\n"
diff --git a/src/main/scala/ifu/icache.scala b/src/main/scala/ifu/icache.scala
index 916a688f..e7181fc1 100644
--- a/src/main/scala/ifu/icache.scala
+++ b/src/main/scala/ifu/icache.scala
@@ -82,6 +82,7 @@ class ICacheBundle(val outer: ICache) extends BoomBundle()(outer.p)
 
   val perf = Output(new Bundle {
     val acquire = Bool()
+    val padd = UInt(paddrBits.W)
   })
 }
 
@@ -334,6 +335,7 @@ class ICacheModule(outer: ICache) extends LazyModuleImp(outer)
   tl_out.e.valid := false.B
 
   io.perf.acquire := tl_out.a.fire
+  io.perf.padd := refill_paddr
 
   when (!refill_valid) { invalidated := false.B }
   when (refill_fire) { refill_valid := true.B }
diff --git a/src/main/scala/lsu/dcache.scala b/src/main/scala/lsu/dcache.scala
index 9f326f50..64fae1b0 100644
--- a/src/main/scala/lsu/dcache.scala
+++ b/src/main/scala/lsu/dcache.scala
@@ -820,6 +820,7 @@ class BoomNonBlockingDCacheModule(outer: BoomNonBlockingDCache) extends LazyModu
 
   io.lsu.perf.release := edge.done(tl_out.c)
   io.lsu.perf.acquire := edge.done(tl_out.a)
+  io.lsu.perf.padd := tl_out.a.bits.address
 
   // load data gen
   val s2_data_word_prebypass = widthMap(w => s2_data_muxed(w) >> Cat(s2_word_idx(w), 0.U(log2Ceil(coreDataBits).W)))
diff --git a/src/main/scala/lsu/lsu.scala b/src/main/scala/lsu/lsu.scala
index 1eb8ba24..ef57c55d 100644
--- a/src/main/scala/lsu/lsu.scala
+++ b/src/main/scala/lsu/lsu.scala
@@ -102,6 +102,7 @@ class LSUDMemIO(implicit p: Parameters, edge: TLEdgeOut) extends BoomBundle()(p)
   val perf = Input(new Bundle {
     val acquire = Bool()
     val release = Bool()
+    val padd = UInt(paddrBits.W)
   })
 
 }
@@ -153,6 +154,11 @@ class LSUCoreIO(implicit p: Parameters) extends BoomBundle()(p)
     val release = Bool()
     val tlbMiss = Bool()
   })
+
+  val tperf = Output(Vec(memWidth, new Bundle {
+    val tlb_miss = Valid(UInt(robAddrSz.W))
+    val cache_miss = Valid(UInt(robAddrSz.W))
+  }))
 }
 
 class LSUIO(implicit p: Parameters, edge: TLEdgeOut) extends BoomBundle()(p)
@@ -1651,6 +1657,32 @@ class LSU(implicit p: Parameters, edge: TLEdgeOut) extends BoomModule()(p)
                     ~(st_exc_killed_mask.asUInt)
 
 
+  // jxy: performance monitor
+  val tlbVadd = RegInit((0.U(vaddrBitsExtended.W)))
+  val cachePadd = RegInit(0.U(coreMaxAddrBits.W))
+  for (w <- 0 until memWidth) {
+    io.core.tperf(w).tlb_miss.valid := false.B
+    io.core.tperf(w).cache_miss.valid := false.B
+    io.core.tperf(w).tlb_miss.bits := DontCare
+    io.core.tperf(w).cache_miss.bits := DontCare
+    when (exe_tlb_valid(w) && (exe_tlb_vaddr(w) >> pgIdxBits) === tlbVadd) {
+      io.core.tperf(w).tlb_miss.valid := true.B
+      io.core.tperf(w).tlb_miss.bits := exe_tlb_uop(w).rob_idx
+      tlbVadd := 0.U
+    }
+    val resp_addr = Mux(
+      io.dmem.resp(w).bits.uop.uses_ldq,
+      ldq(io.dmem.resp(w).bits.uop.ldq_idx).bits.addr.bits,
+      stq(io.dmem.resp(w).bits.uop.stq_idx).bits.addr.bits
+    )
+    when (io.dmem.resp(w).valid && (resp_addr >> blockOffBits) === (cachePadd >> blockOffBits)) {
+      io.core.tperf(w).cache_miss.valid := true.B
+      io.core.tperf(w).cache_miss.bits := io.dmem.resp(w).bits.uop.rob_idx
+      cachePadd := 0.U
+    }
+  }
+  when (dtlb.io.perf.miss)    { tlbVadd   := dtlb.io.perf.vadd }
+  when (io.dmem.perf.acquire) { cachePadd := io.dmem.perf.padd }
 }
 
 /**
diff --git a/src/main/scala/lsu/tlb.scala b/src/main/scala/lsu/tlb.scala
index a09018b4..311b0412 100644
--- a/src/main/scala/lsu/tlb.scala
+++ b/src/main/scala/lsu/tlb.scala
@@ -23,6 +23,11 @@ class NBDTLB(instruction: Boolean, lgMaxSize: Int, cfg: TLBConfig)(implicit edge
     val sfence = Input(Valid(new SFenceReq))
     val ptw = new TLBPTWIO
     val kill = Input(Bool())
+    // jxy: performance monitor
+    val perf = Output(new Bundle {
+      val miss = Output(Bool())
+      val vadd = Output(UInt(vaddrBitsExtended.W))
+    })
   })
   io.ptw := DontCare
   io.resp := DontCare
@@ -358,4 +363,7 @@ class NBDTLB(instruction: Boolean, lgMaxSize: Int, cfg: TLBConfig)(implicit edge
   }
 
 
+  // jxy: performance monitor
+  io.perf.miss := io.ptw.req.fire
+  io.perf.vadd := io.ptw.req.bits.bits.addr
 }
