diff --git a/src/main/scala/common/config-mixins.scala b/src/main/scala/common/config-mixins.scala
index 5de8ebf4..b716c555 100644
--- a/src/main/scala/common/config-mixins.scala
+++ b/src/main/scala/common/config-mixins.scala
@@ -156,7 +156,7 @@ class WithNMediumBooms(n: Int = 1, overrideIdOffset: Option[Int] = None) extends
               fpu = Some(freechips.rocketchip.tile.FPUParams(sfmaLatency=4, dfmaLatency=4, divSqrt=true))
             ),
             dcache = Some(
-              DCacheParams(rowBits = 64, nSets=64, nWays=4, nMSHRs=2, nTLBWays=8)
+              DCacheParams(rowBits = 64, nSets=64, nWays=4, nMSHRs=2, nTLBWays=16)
             ),
             icache = Some(
               ICacheParams(rowBits = 64, nSets=64, nWays=4, fetchBytes=2*4)
@@ -437,23 +437,25 @@ class WithNCS152DefaultBooms(n: Int = 1, overrideIdOffset: Option[Int] = None) e
 class WithTAGELBPD extends Config((site, here, up) => {
   case TilesLocated(InSubsystem) => up(TilesLocated(InSubsystem), site) map {
     case tp: BoomTileAttachParams => tp.copy(tileParams = tp.tileParams.copy(core = tp.tileParams.core.copy(
-      bpdMaxMetaLength = 120,
+      bpdMaxMetaLength = 400,
       globalHistoryLength = 64,
       localHistoryLength = 1,
       localHistoryNSets = 0,
       branchPredictor = ((resp_in: BranchPredictionBankResponse, p: Parameters) => {
         val loop = Module(new LoopBranchPredictorBank()(p))
         val tage = Module(new TageBranchPredictorBank()(p))
+        val ibp = Module(new IndirectBranchPredictorBank()(p))
         val btb = Module(new BTBBranchPredictorBank()(p))
         val bim = Module(new BIMBranchPredictorBank()(p))
         val ubtb = Module(new FAMicroBTBBranchPredictorBank()(p))
-        val preds = Seq(loop, tage, btb, ubtb, bim)
+        val preds = Seq(loop, tage, ibp, btb, ubtb, bim)
         preds.map(_.io := DontCare)
 
         ubtb.io.resp_in(0)  := resp_in
         bim.io.resp_in(0)   := ubtb.io.resp
         btb.io.resp_in(0)   := bim.io.resp
-        tage.io.resp_in(0)  := btb.io.resp
+        ibp.io.resp_in(0)   := btb.io.resp
+        tage.io.resp_in(0)  := ibp.io.resp
         loop.io.resp_in(0)  := tage.io.resp
 
         (preds, loop.io.resp)
diff --git a/src/main/scala/common/parameters.scala b/src/main/scala/common/parameters.scala
index 47da9ee5..888fe364 100644
--- a/src/main/scala/common/parameters.scala
+++ b/src/main/scala/common/parameters.scala
@@ -65,6 +65,10 @@ case class BoomCoreParams(
   bpdMaxMetaLength: Int = 120,
   numRasEntries: Int = 32,
   enableRasTopRepair: Boolean = true,
+  indirectHistoryLength: Int = 16,
+  indirectHistoryWidth: Int = 16,
+  loadHistoryLength: Int = 16,
+  loadHistoryWidth: Int = 16,
 
   /* more stuff */
   useCompressed: Boolean = true,
@@ -257,6 +261,11 @@ trait HasBoomCoreParameters extends freechips.rocketchip.tile.HasCoreParameters
 
   val useLHist = localHistoryNSets > 1 && localHistoryLength > 1
 
+  val indirectHistoryLength = boomParams.indirectHistoryLength
+  val indirectHistoryWidth = boomParams.indirectHistoryWidth
+  val loadHistoryLength = boomParams.loadHistoryLength
+  val loadHistoryWidth = boomParams.loadHistoryWidth
+
   //************************************
   // Extra Knobs and Features
   val enableCommitMapTable = boomParams.enableCommitMapTable
diff --git a/src/main/scala/common/tile.scala b/src/main/scala/common/tile.scala
index 05821c52..dd1c9032 100644
--- a/src/main/scala/common/tile.scala
+++ b/src/main/scala/common/tile.scala
@@ -6,7 +6,7 @@
 package boom.common
 
 import chisel3._
-import chisel3.util.{RRArbiter, Queue}
+import chisel3.util._
 
 import scala.collection.mutable.{ListBuffer}
 
@@ -243,6 +243,9 @@ class BoomTileModuleImp(outer: BoomTile) extends BaseTileModuleImp(outer){
   lsu.io.hellacache <> hellaCacheArb.io.mem
   outer.dcache.module.io.lsu <> lsu.io.dmem
 
+  // External performance monitor
+  outer.pmSourceNode.bundle <> core.io.pm
+
   // Generate a descriptive string
   val frontendStr = outer.frontend.module.toString
   val coreStr = core.toString
diff --git a/src/main/scala/exu/core.scala b/src/main/scala/exu/core.scala
index 0f8d39e3..4ab314aa 100644
--- a/src/main/scala/exu/core.scala
+++ b/src/main/scala/exu/core.scala
@@ -37,7 +37,7 @@ import org.chipsalliance.cde.config.Parameters
 import freechips.rocketchip.rocket.Instructions._
 import freechips.rocketchip.tile.{TraceBundle}
 import freechips.rocketchip.rocket.{Causes, PRV, TracedInstruction}
-import freechips.rocketchip.util.{Str, UIntIsOneOf, CoreMonitorBundle}
+import freechips.rocketchip.util.{Str, UIntIsOneOf, CoreMonitorBundle, PmSignals}
 import freechips.rocketchip.devices.tilelink.{PLICConsts, CLINTConsts}
 
 import boom.common._
@@ -61,6 +61,7 @@ class BoomCore()(implicit p: Parameters) extends BoomModule
     val ptw_tlb = new freechips.rocketchip.rocket.TLBPTWIO()
     val trace = Output(new TraceBundle)
     val fcsr_rm = UInt(freechips.rocketchip.tile.FPConstants.RM_SZ.W)
+    val pm = Output(new PmSignals)
   })
 
   io.ptw_tlb := DontCare
@@ -241,6 +242,13 @@ class BoomCore()(implicit p: Parameters) extends BoomModule
     mem_units(i).io.lsu_io <> io.lsu.exe(i)
   }
 
+  val load_rob_idx_history = Reg(Vec(loadHistoryLength, UInt(robAddrSz.W)))
+  val load_pc_history = Reg(Vec(loadHistoryLength, UInt(loadHistoryWidth.W)))
+  val load_val_history = Reg(Vec(loadHistoryLength, UInt(loadHistoryWidth.W)))
+
+  io.ifu.load_pc_history := load_pc_history.asUInt
+  io.ifu.load_val_history := load_val_history.asUInt
+
   //-------------------------------------------------------------
   // Uarch Hardware Performance Events (HPEs)
 
@@ -448,7 +456,9 @@ class BoomCore()(implicit p: Parameters) extends BoomModule
       true.B,
       io.ifu.get_pc(1).pc,
       ftq_entry.cfi_is_call && ftq_entry.cfi_idx.bits === cfi_idx,
-      ftq_entry.cfi_is_ret  && ftq_entry.cfi_idx.bits === cfi_idx)
+      ftq_entry.cfi_is_ret  && ftq_entry.cfi_idx.bits === cfi_idx,
+      ftq_entry.cfi_type === CFI_JALR && !ftq_entry.cfi_is_ret && ftq_entry.cfi_idx.bits === cfi_idx,
+      brupdate.b2.jalr_target)
 
 
     io.ifu.redirect_ghist   := Mux(
@@ -515,6 +525,44 @@ class BoomCore()(implicit p: Parameters) extends BoomModule
     dec_uops(w) := decode_units(w).io.deq.uop
   }
 
+  val next_load_rob_idx_history = WireInit(load_rob_idx_history)
+  val next_load_pc_history = WireInit(load_pc_history)
+  val next_load_val_history = WireInit(load_val_history)
+  val is_load = Wire(Vec(coreWidth, Bool()))
+  for (w <- 0 until coreWidth) {
+    is_load(w) := dis_valids(w) && dis_uops(w).uopc === uopLD
+  }
+  val prefix_sum = Wire(Vec(coreWidth+1, UInt(log2Ceil(coreWidth+1).W)))
+  prefix_sum(0) := 0.U
+  for (w <- 1 until coreWidth + 1) {
+    prefix_sum(w) := prefix_sum(w-1) + is_load(w-1)
+  }
+  next_load_rob_idx_history :=
+    (load_rob_idx_history.asUInt << PopCount(is_load.asUInt) * robAddrSz.U).asTypeOf(load_rob_idx_history)
+  next_load_pc_history :=
+    (load_pc_history.asUInt << PopCount(is_load.asUInt) * loadHistoryWidth.U).asTypeOf(load_pc_history)
+  next_load_val_history :=
+    (load_val_history.asUInt << PopCount(is_load.asUInt) * loadHistoryWidth.U).asTypeOf(load_val_history)
+  for (w <- 0 until coreWidth) {
+    when (is_load(w)) {
+      next_load_rob_idx_history(prefix_sum(coreWidth) - 1.U - prefix_sum(w)) := dis_uops(w).rob_idx
+      next_load_pc_history(prefix_sum(coreWidth) - 1.U - prefix_sum(w)) := dis_uops(w).debug_pc
+      next_load_val_history(prefix_sum(coreWidth) - 1.U - prefix_sum(w)) := 0.U
+    }
+  }
+  for (i <- 0 until memWidth) {
+    when (mem_resps(i).valid) {
+      for (j <- 0 until loadHistoryLength) {
+        when (load_rob_idx_history(j) === mem_resps(i).bits.uop.rob_idx) {
+          next_load_val_history(j.U+PopCount(is_load.asUInt)) := mem_resps(i).bits.data
+        }
+      }
+    }
+  }
+  load_rob_idx_history := next_load_rob_idx_history
+  load_pc_history := next_load_pc_history
+  load_val_history := next_load_val_history
+
   //-------------------------------------------------------------
   // FTQ GetPC Port Arbitration
 
@@ -1492,4 +1540,226 @@ class BoomCore()(implicit p: Parameters) extends BoomModule
   } else {
     io.ifu.debug_ftq_idx := DontCare
   }
+
+  // jxy: performance monitor
+  val maskWidth = 4
+  val pidWidth = 11
+  val evtWidth = 3
+  val stageWidth = 2
+  val cntWidth = 4
+  val pipLatency = 14.U
+  val level = Wire(UInt(2.W)) // global
+  val pid   = RegInit(0.U(pidWidth.W))
+  val pc    = Wire(Vec(coreWidth, UInt(vaddrBits.W))) // per instruction
+  val addr  = Wire(Vec(coreWidth, UInt(vaddrBits.W)))
+  val stage = Wire(Vec(coreWidth, UInt(stageWidth.W)))
+  val evt   = RegInit(VecInit(Seq.fill(coreWidth)(0.U(evtWidth.W))))
+  val evm   = Wire(Vec(coreWidth, Vec(maskWidth, Bool())))
+  for (w <- 0 until coreWidth) {
+    io.pm.tag(w)   := RegNext(Cat(pid, level, evt(w)))
+    io.pm.info(w)  := RegNext(Cat(addr(w), stage(w), pc(w)))
+    io.pm.event(w) := RegNext(evm(w).asUInt)
+  }
+  // Additional info in ROB
+  val next_pc_reg = RegInit(0.U(vaddrBits.W))
+  val last_pc_reg = RegInit(0.U(vaddrBits.W))
+  val last_cnt_reg = RegInit(0.U(cntWidth.W))
+  val last_evm_reg = RegInit(VecInit(Seq.fill(maskWidth)(false.B)))
+  val next_pc = Wire(UInt(vaddrBits.W))
+  val last_pc = Wire(UInt(vaddrBits.W))
+  val last_cnt = Wire(UInt(cntWidth.W))
+  val last_evm = Wire(Vec(maskWidth, Bool()))
+  val next_evm = Wire(Vec(maskWidth, Bool()))
+  val next_stage = Wire(UInt(stageWidth.W))
+  next_pc_reg := next_pc
+  last_pc_reg := last_pc
+  last_cnt_reg := last_cnt
+  last_evm_reg := last_evm
+  next_pc := next_pc_reg
+  last_pc := last_pc_reg
+  last_cnt := Mux(last_cnt_reg =/= 0.U && rob.io.ready, last_cnt_reg - 1.U, last_cnt_reg)
+  last_evm := last_evm_reg
+  next_stage := 0.U
+  next_evm := last_evm
+  for (w <- 0 until coreWidth) {
+    def GetRowIdx(rob_idx: UInt): UInt = {
+      if (coreWidth == 1) return rob_idx
+      else return rob_idx >> log2Ceil(coreWidth).U
+    }
+    def GetBankIdx(rob_idx: UInt): UInt = {
+      if (coreWidth == 1) return 0.U
+      else return rob_idx(log2Ceil(coreWidth)-1, 0).asUInt
+    }
+    def MatchBank(bank_idx: UInt): Bool = (bank_idx === w.U)
+    val rob_stage = RegInit(VecInit(Seq.fill(numRobRows)(0.U(stageWidth.W))))
+    val rob_pid = RegInit(VecInit(Seq.fill(numRobRows)(0.U(pidWidth.W))))
+    val rob_misp = RegInit(VecInit(Seq.fill(numRobRows)(false.B)))
+    val rob_taken = RegInit(VecInit(Seq.fill(numRobRows)(false.B)))
+    val rob_addr = RegInit(VecInit(Seq.fill(numRobRows)(0.U(vaddrBits.W))))
+    val rob_dtm = RegInit(VecInit(Seq.fill(numRobRows)(false.B)))
+    val rob_dcm = RegInit(VecInit(Seq.fill(numRobRows)(false.B)))
+    val rob_cnt = RegInit(VecInit(Seq.fill(numRobRows)(0.U(cntWidth.W))))
+    val com_pc = Wire(UInt(vaddrBits.W))
+    val com_cnt = Wire(UInt(cntWidth.W))
+    val com_evm = Wire(Vec(maskWidth, Bool()))
+    val com_addr = Wire(UInt(vaddrBits.W))
+    val com_evt = Wire(UInt(evtWidth.W))
+    // Count for redirect scope
+    for (i <- 0 until numRobRows) {
+      when (rob_cnt(i) =/= 0.U && rob.io.ready) {
+        rob_cnt(i) := rob_cnt(i) - 1.U
+      }
+    }
+    // Dispatching
+    for (i <- 0 until coreWidth) {
+      when(dis_valids(i) && MatchBank(GetBankIdx(dis_uops(i).rob_idx))) {
+        rob_stage(GetRowIdx(dis_uops(i).rob_idx)) := 1.U
+        rob_cnt(GetRowIdx(dis_uops(i).rob_idx)) := 0.U
+        rob_misp(GetRowIdx(dis_uops(i).rob_idx)) := false.B
+        rob_dtm(GetRowIdx(dis_uops(i).rob_idx)) := false.B
+        rob_dcm(GetRowIdx(dis_uops(i).rob_idx)) := false.B
+      }
+    }
+    // Issuing
+    for (i <- 0 until exe_units.length) {
+      when (iss_valids(i) && MatchBank(GetBankIdx(iss_uops(i).rob_idx))) {
+        rob_stage(GetRowIdx(iss_uops(i).rob_idx)) := 2.U
+      }
+    }
+    // Record PID from HINT argument
+    for (i <- 0 until rob.numWakeupPorts) {
+      val wb_resp = rob.io.wb_resps(i)
+      val wb_uop = wb_resp.bits.uop
+      when (wb_resp.valid && MatchBank(GetBankIdx(wb_uop.rob_idx))) {
+        rob_stage(GetRowIdx(wb_uop.rob_idx)) := 3.U
+        rob_pid(GetRowIdx(wb_uop.rob_idx)) := wb_resp.bits.data.asUInt
+      }
+    }
+    // Record branches and PCs
+    for (b <- brinfos) {
+      when (b.valid && MatchBank(GetBankIdx(b.uop.rob_idx))) {
+        rob_cnt(GetRowIdx(b.uop.rob_idx)) := Mux(b.mispredict, pipLatency, 0.U)
+        rob_misp(GetRowIdx(b.uop.rob_idx)) := b.mispredict
+        rob_taken(GetRowIdx(b.uop.rob_idx)) := b.taken
+        rob_addr(GetRowIdx(b.uop.rob_idx)) := b.jalr_target
+      }
+    }
+    // Record memory instruction addresses
+    for (i <- 0 until memWidth) {
+      when (io.lsu.exe(i).req.valid) {
+        when (MatchBank(GetBankIdx(io.lsu.exe(i).req.bits.uop.rob_idx))) {
+          rob_addr(GetRowIdx(io.lsu.exe(i).req.bits.uop.rob_idx)) := io.lsu.exe(i).req.bits.addr
+        }
+      }
+    }
+    // Record DTLB and data cache misses
+    for (i <- 0 until memWidth) {
+      when (io.lsu.tperf(i).tlb_miss.valid && MatchBank(GetBankIdx(io.lsu.tperf(i).tlb_miss.bits))) {
+        rob_dtm(GetRowIdx(io.lsu.tperf(i).tlb_miss.bits)) := true.B
+      }
+      when (io.lsu.tperf(i).cache_miss.valid && MatchBank(GetBankIdx(io.lsu.tperf(i).cache_miss.bits))) {
+        rob_dcm(GetRowIdx(io.lsu.tperf(i).cache_miss.bits)) := true.B
+      }
+    }
+    // Record ITLB and instruction cache miss
+    val itm = Wire(Bool())
+    val icm = Wire(Bool())
+    itm := io.ifu.tperf.tlb_miss(rob.io.commit.uops(w).ftq_idx)
+    icm := io.ifu.tperf.cache_miss(rob.io.commit.uops(w).ftq_idx)
+    for (i <- 0 until w) {
+      when (rob.io.commit.uops(w).ftq_idx === rob.io.commit.uops(i).ftq_idx) {
+        itm := false.B
+        icm := false.B
+      }
+    }
+    // Committing
+    when (rob.io.commit.valids(w)) {
+      rob_stage(GetRowIdx(rob.io.commit.uops(w).rob_idx)) := 0.U
+      rob_misp(GetRowIdx(rob.io.commit.uops(w).rob_idx)) := false.B
+      rob_dtm(GetRowIdx(rob.io.commit.uops(w).rob_idx)) := false.B
+      rob_dcm(GetRowIdx(rob.io.commit.uops(w).rob_idx)) := false.B
+    }
+    val uop = RegNext(rob.io.commit.uops(w))
+    when (RegNext(rob.io.commit.valids(w))) {
+      last_evm := RegNext(com_evm)
+      last_pc := com_pc
+      last_cnt := com_cnt
+      when (uop.is_br && rob_taken(GetRowIdx(uop.rob_idx))) {
+        next_pc := (com_pc.asSInt + ImmGen(uop.imm_packed, IS_B)).asUInt
+      }.elsewhen (uop.is_jal) {
+        next_pc := (com_pc.asSInt + ImmGen(uop.imm_packed, IS_J)).asUInt
+      }.elsewhen (uop.is_jalr) {
+        next_pc := rob_addr(GetRowIdx(uop.rob_idx))
+      }.otherwise {
+        next_pc := com_pc + Mux(uop.is_rvc, 2.U, 4.U)
+      }
+    }
+    when (RegNext(rob.io.flush.valid)) {
+      last_cnt := pipLatency
+      next_pc := io.ifu.redirect_pc
+    }
+    when (MatchBank(GetBankIdx(rob.io.rob_head_idx))) {
+      next_stage := Mux(rob.io.empty, 0.U, rob_stage(GetRowIdx(rob.io.rob_head_idx)))
+      next_evm := Mux(rob.io.empty, VecInit(Seq.fill(maskWidth)(false.B)), com_evm)
+      next_evm(0) := 1.B
+    }
+    // Get stage, event mask and PC
+    io.ifu.debug_ftq_idx(w) := rob.io.commit.uops(w).ftq_idx
+    com_pc := (AlignPCToBoundary(io.ifu.debug_fetch_pc(w), icBlockBytes)
+      + uop.pc_lob - Mux(uop.edge_inst, 2.U, 0.U))(vaddrBits-1,0)
+    com_cnt := rob_cnt(GetRowIdx(uop.rob_idx))
+    stage(w) := 0.U
+    com_evm(0) := 1.B
+    com_evm(1) := itm || rob_dtm(GetRowIdx(rob.io.commit.uops(w).rob_idx))
+    com_evm(2) := icm
+    com_evm(3) :=
+      rob_dcm(GetRowIdx(rob.io.commit.uops(w).rob_idx)) ||
+      rob_misp(GetRowIdx(rob.io.commit.uops(w).rob_idx))
+    com_addr := rob_addr(GetRowIdx(rob.io.commit.uops(w).rob_idx))
+    when (evt(w) === 0.U) {
+      when (last_cnt === 0.U) {
+        pc(w) := next_pc
+        evm(w) := next_evm
+        stage(w) := next_stage
+        addr(w) := 0.U
+      }.otherwise {
+        pc(w) := last_pc
+        evm(w) := last_evm
+        addr(w) := 0.U
+      }
+    }.otherwise {
+      pc(w) := com_pc
+      evm(w) := RegNext(com_evm)
+      addr(w) := RegNext(com_addr)
+    }
+    // Get PID
+    when (
+      rob.io.commit.valids(w) &&
+      rob.io.commit.uops(w).uopc === uopSLLI &&
+      rob.io.commit.uops(w).pdst === 0.U
+    ) {
+      pid := rob_pid(GetRowIdx(rob.io.commit.uops(w).rob_idx))
+    }
+    // Get event tag
+    when (rob.io.commit.uops(w).is_jalr && rob.io.commit.uops(w).ldst === 0.U && (
+      rob.io.commit.uops(w).lrs1 === 1.U || rob.io.commit.uops(w).lrs1 === 5.U
+    )) {
+      com_evt := 1.U // EVT::RET
+    }.elsewhen (rob.io.commit.uops(w).is_jalr) {
+      com_evt := 2.U // EVT:JALR
+    }.elsewhen (rob.io.commit.uops(w).is_br) {
+      com_evt := 3.U // EVT:BRANCH
+    }.elsewhen (rob.io.commit.uops(w).is_jal) {
+      com_evt := 4.U // EVT:JAL
+    }.otherwise {
+      com_evt := 5.U // EVT::OTHERS
+    }
+    when (!rob.io.commit.valids(w)) {
+      evt(w) := 0.U // EVT::NULL
+    }.otherwise {
+      evt(w) := com_evt
+    }
+  }
+  // Get privilege level
+  level := csr.io.status.prv
 }
diff --git a/src/main/scala/ifu/bpd/composer.scala b/src/main/scala/ifu/bpd/composer.scala
index 997c9b62..24a1dc79 100644
--- a/src/main/scala/ifu/bpd/composer.scala
+++ b/src/main/scala/ifu/bpd/composer.scala
@@ -26,6 +26,9 @@ class ComposedBranchPredictorBank(implicit p: Parameters) extends BranchPredicto
     c.io.f0_mask   := io.f0_mask
     c.io.f1_ghist  := io.f1_ghist
     c.io.f1_lhist  := io.f1_lhist
+    c.io.f1_phist  := io.f1_phist
+    c.io.f1_lphist  := io.f1_lphist
+    c.io.f1_lvhist  := io.f1_lvhist
     c.io.f3_fire   := io.f3_fire
     if (c.metaSz > 0) {
       metas = (metas << c.metaSz) | c.io.f3_meta(c.metaSz-1,0)
diff --git a/src/main/scala/ifu/bpd/ibp.scala b/src/main/scala/ifu/bpd/ibp.scala
new file mode 100644
index 00000000..b5712c82
--- /dev/null
+++ b/src/main/scala/ifu/bpd/ibp.scala
@@ -0,0 +1,424 @@
+package boom.ifu
+
+import chisel3._
+import chisel3.util._
+
+import org.chipsalliance.cde.config.{Field, Parameters}
+import freechips.rocketchip.diplomacy._
+import freechips.rocketchip.tilelink._
+
+import boom.common._
+import boom.util.{BoomCoreStringPrefix}
+
+import scala.math.min
+
+case class BoomIndirectParams(
+  // bhLen, phLen, nSets, nWays
+  baseTableInfo: Seq[(Int, Int, Int, Int)] = Seq(
+    (4,  1,  256, 1),
+    (8,  2,  256, 1),
+    (16, 4, 1024, 1),
+    (32, 8,  512, 1)
+  ),
+  // nSets, nWays
+  lvcTableInfo: Seq[(Int, Int)] = Seq.fill(16)((16, 2)),
+  tagSz: Int = 16,
+  maxWays: Int = 8,
+  uBitPeriod: Int = 2048,
+  selSets: Int = 64,
+  selBits: Int = 6,
+  selPeriod: Int = 1048576,
+  lvcSets: Int = 256,
+  lvcCntBits: Int = 3,
+  metaRespBits: Int = 8
+)
+
+class IndirectBranchPredictionTable(
+  override val nSets: Int, override val nWays: Int, val tagSz: Int, val metaSz: Int
+)(implicit p: Parameters) extends BoomModule()(p) with HasBoomFrontendParameters
+{
+  val io = IO(new Bundle {
+    val s1_valid          = Input(Bool())
+    val s1_tag            = Input(UInt(tagSz.W))
+    val update_valid_meta = Input(Bool())
+    val update_valid_tag  = Input(Bool())
+    val update_cfi_idx    = Input(UInt(log2Ceil(bankWidth).W))
+    val update_meta       = Input(UInt(metaSz.W))
+    val update_plru       = Input(UInt(nWays.W))
+    val update_tag        = Input(UInt(tagSz.W))
+    val update_way        = Input(UInt(log2Ceil(nWays).W))
+    val update_target     = Input(UInt(vaddrBitsExtended.W))
+    val s2_resp           = Output(Vec(bankWidth, Valid(UInt(vaddrBitsExtended.W))))
+    val s2_meta           = Output(Vec(bankWidth, UInt(metaSz.W)))
+    val s2_plru           = Output(Vec(bankWidth, UInt(nWays.W)))
+    val s2_tag            = Output(UInt(tagSz.W))
+    val s2_way            = Output(Vec(bankWidth, UInt(log2Ceil(nWays).W)))
+  })
+
+  def foldTag2Idx(bits: UInt): UInt = {
+    ((0 until (tagSz + log2Ceil(nSets) - 1) / log2Ceil(nSets)) map { i =>
+      bits(min((i + 1) * log2Ceil(nSets), tagSz) - 1, i * log2Ceil(nSets))
+    }).reduce(_^_)
+    // bits(log2Ceil(nSets) - 1, 0) // uncomment for verification with no folding
+  }
+
+  def pLRU(bdt: UInt, nWays: Int): UInt = {
+    val bdt_pos = Wire(Vec(log2Ceil(nWays), Bool()))
+    var bdt_cur = 0.U
+    for (i <- log2Ceil(nWays) - 1 to 0 by -1) {
+      bdt_pos(i) := bdt(bdt_cur)
+      bdt_cur = (bdt_cur << 1) + Mux(bdt(bdt_cur), 2.U, 1.U)
+    }
+    if (nWays > 1) bdt_pos.asUInt else 0.U
+  }
+
+  def pLRUNext(bdt: UInt, access: UInt, nWays: Int): UInt = {
+    val new_bdt = Wire(Vec(nWays, Bool()))
+    var bdt_cur = 0.U
+    new_bdt := bdt.asBools
+    for (i <- log2Ceil(nWays) - 1 to 0 by -1) {
+      new_bdt(bdt_cur) := ~access(i)
+      bdt_cur = (bdt_cur << 1) + Mux(access(i), 2.U, 1.U)
+    }
+    new_bdt.asUInt
+  }
+
+  // reset
+  val doing_reset = RegInit(true.B)
+  val reset_idx   = RegInit(0.U(log2Ceil(nSets).W))
+  reset_idx := reset_idx + doing_reset
+  when (reset_idx === (nSets-1).U) { doing_reset := false.B }
+
+  // memory structure
+  val tags = Seq.fill(nWays) { SyncReadMem(nSets, Vec(bankWidth, UInt(tagSz.W))) }
+  val ibtb = Seq.fill(nWays) { SyncReadMem(nSets, Vec(bankWidth, UInt(vaddrBitsExtended.W))) }
+  val meta = Seq.fill(nWays) { SyncReadMem(nSets, Vec(bankWidth, UInt(metaSz.W))) }
+  val plru = SyncReadMem(nSets, Vec(bankWidth, UInt(nWays.W)))
+
+  // read/write ports and data
+  val s1_req_idx = foldTag2Idx(io.s1_tag)
+  val s2_req_tag = RegNext(io.s1_tag)
+  val s2_req_ribtb = VecInit(ibtb.map { b =>
+    VecInit(b.read(s1_req_idx, io.s1_valid).map(_.asTypeOf(UInt(vaddrBitsExtended.W))))
+  })
+  val s2_req_rtags = VecInit(tags.map { t =>
+    VecInit(t.read(s1_req_idx, io.s1_valid).map(_.asTypeOf(UInt(tagSz.W))))
+  })
+  val s2_req_rmeta = VecInit(meta.map { u =>
+    VecInit(u.read(s1_req_idx, io.s1_valid).map(_.asTypeOf(UInt(metaSz.W))))
+  })
+  val s2_req_rplru = VecInit(plru.read(s1_req_idx, io.s1_valid).map(_.asTypeOf(UInt(nWays.W))))
+
+  // tag matching and output results
+  val s2_hit_ohs = VecInit((0 until bankWidth) map { i =>
+    VecInit((0 until nWays) map { w =>
+      s2_req_rtags(w)(i) === s2_req_tag
+    })
+  })
+  val s2_hits     = s2_hit_ohs.map { oh => oh.reduce(_||_) }
+  val s2_hit_ways = s2_hit_ohs.map { oh => PriorityEncoder(oh) }
+  for (w <- 0 until bankWidth) {
+    val access_way = Mux(s2_hits(w), s2_hit_ways(w), pLRU(s2_req_rplru(w), nWays))
+    io.s2_resp(w).valid := !doing_reset && RegNext(io.s1_valid) && s2_hits(w)
+    io.s2_resp(w).bits  := s2_req_ribtb(s2_hit_ways(w))(w)
+    io.s2_meta(w)       := s2_req_rmeta(s2_hit_ways(w))(w)
+    io.s2_plru(w)       := pLRUNext(s2_req_rplru(w), access_way, nWays)
+    io.s2_way (w)       := access_way
+  }
+  io.s2_tag := s2_req_tag
+
+  // update entry when mispredicted
+  val update_wtag_mask = (
+    UIntToOH(io.update_cfi_idx) &
+    Fill(bankWidth, io.update_valid_tag)
+  )
+  val update_wmeta_mask = (
+    UIntToOH(io.update_cfi_idx) &
+    Fill(bankWidth, io.update_valid_meta)
+  )
+  for (w <- 0 until nWays) {
+    when (doing_reset || io.update_way === w.U || (w == 0 && nWays == 1).B) {
+      // write variables
+      val widx      = Mux(doing_reset, reset_idx, foldTag2Idx(io.update_tag))
+      val tags_mask = Mux(doing_reset, ~(0.U(bankWidth.W)), update_wtag_mask).asBools
+      val meta_mask = Mux(doing_reset, ~(0.U(bankWidth.W)), update_wmeta_mask).asBools
+      val ibtb_wdat = Mux(
+        doing_reset,
+        VecInit(Seq.fill(bankWidth) { 0.U(vaddrBitsExtended.W) }),
+        VecInit(Seq.fill(bankWidth) { io.update_target })
+      )
+      val tags_wdat = Mux(
+        doing_reset,
+        VecInit(Seq.fill(bankWidth) { 0.U(tagSz.W) }),
+        VecInit(Seq.fill(bankWidth) { io.update_tag })
+      )
+      val meta_wdat = Mux(
+        doing_reset,
+        VecInit(Seq.fill(bankWidth) { 0.U(metaSz.W) }),
+        VecInit(Seq.fill(bankWidth) { io.update_meta })
+      )
+      val plru_wdat = Mux(
+        doing_reset,
+        VecInit(Seq.fill(bankWidth) { 0.U(nWays.W) }),
+        VecInit(Seq.fill(bankWidth) { io.update_plru })
+      )
+      // write entry
+      tags(w).write(widx, tags_wdat, tags_mask)
+      ibtb(w).write(widx, ibtb_wdat, tags_mask)
+      meta(w).write(widx, meta_wdat, meta_mask)
+      plru   .write(widx, plru_wdat, meta_mask)
+    }
+  }
+}
+
+class IndirectBranchPredictorBank(params: BoomIndirectParams = BoomIndirectParams())(implicit p: Parameters)
+  extends BranchPredictorBank()(p)
+{
+  class IBTBPredictMeta extends Bundle {
+    // base predictor meta
+    val base_provider = Vec(bankWidth, Valid(UInt(log2Ceil(params.baseTableInfo.length+1).W)))
+    val base_allocate = Vec(bankWidth, Valid(UInt(log2Ceil(params.baseTableInfo.length+1).W)))
+    val base_hashed   = Vec(bankWidth, Output(UInt(params.metaRespBits.W)))
+    val base_alt_diff = Vec(bankWidth, Output(Bool()))
+    val base_useful   = Vec(bankWidth, Output(Bool()))
+    val base_plru     = Vec(bankWidth, Output(UInt(params.maxWays.W)))
+    val base_prv_tag  = Output(UInt(params.tagSz.W))
+    val base_alc_tag  = Output(UInt(params.tagSz.W))
+    val base_prv_way  = Vec(bankWidth, Output(UInt(log2Ceil(params.maxWays).W)))
+    val base_alc_way  = Vec(bankWidth, Output(UInt(log2Ceil(params.maxWays).W)))
+    // lvc predictor meta
+    val lvc_provider  = Vec(bankWidth, Output(UInt(log2Ceil(params.lvcTableInfo.length).W)))
+    val lvc_hashed    = Vec(bankWidth, Output(UInt(params.metaRespBits.W)))
+    val lvc_cnt       = Vec(bankWidth, Output(UInt(params.lvcCntBits.W)))
+    val lvc_plru      = Vec(bankWidth, Output(UInt(params.maxWays.W)))
+    val lvc_prv_tag   = Output(UInt(params.tagSz.W))
+    val lvc_prv_way   = Vec(bankWidth, Output(UInt(log2Ceil(params.maxWays).W)))
+  }
+
+  def foldMetaResp(bits: UInt): UInt = {
+    ((0 until (vaddrBitsExtended + params.metaRespBits - 1) / params.metaRespBits) map {
+      i => bits(min((i + 1) * params.metaRespBits, vaddrBitsExtended) - 1, i * params.metaRespBits)
+    }).reduce(_^_)
+  }
+
+  // global update signals
+  val update_valid =
+    s1_update.valid &&                 // valid data
+    s1_update.bits.cfi_idx.valid &&    // a CFI instruction
+    s1_update.bits.is_commit_update && // update on commit
+    s1_update.bits.cfi_is_jalr         // a JALR instruction
+  val update_meta = s1_update.bits.meta.asTypeOf(new IBTBPredictMeta)
+  val update_idx = s1_update.bits.cfi_idx.bits
+
+  // tournament selection counters
+  val sel = RegInit(VecInit(Seq.fill(params.selSets)(0.U(params.selBits.W))))
+  val update_sel_old = sel(fetchIdx(s1_update.bits.pc))
+  val update_sel_new = WireInit(update_sel_old)
+  val update_folded_target = foldMetaResp(s1_update.bits.target)
+  when (
+    update_meta.lvc_hashed (update_idx) === update_folded_target &&
+    update_meta.base_hashed(update_idx) =/= update_folded_target
+  ) {
+    val saturated = (1.U << params.selBits) - 1.U
+    update_sel_new := Mux(update_sel_old === saturated, saturated, update_sel_old + 1.U)
+  }.elsewhen (
+    update_meta.lvc_hashed (update_idx) =/= update_folded_target &&
+    update_meta.base_hashed(update_idx) === update_folded_target
+  ) {
+    update_sel_new := Mux(update_sel_old === 0.U, 0.U, update_sel_old - 1.U)
+  }
+  when (update_valid) {
+    sel(fetchIdx(s1_update.bits.pc)) := update_sel_new
+  }
+
+  // responses
+  val s2_lvcsel = RegNext(sel(fetchIdx(s1_pc))(params.selBits-1))
+  val s2_resp = Wire(Vec(bankWidth, Valid(UInt(vaddrBitsExtended.W))))
+  val s2_meta = WireInit(0.U.asTypeOf(new IBTBPredictMeta))
+
+  // base predictor
+  val baseP = params.baseTableInfo.zipWithIndex.map { case ((bhLen, phLen, nSets, nWays), i) =>
+    val t = Module(new IndirectBranchPredictionTable(nSets, nWays, params.tagSz, 1))
+    // calculate folded path history
+    require(bhLen        <= globalHistoryLength)
+    require(phLen        <= indirectHistoryLength)
+    require(params.tagSz <= indirectHistoryWidth)
+    val folded_ghist = if (bhLen > 0) {
+      ((0 until (bhLen + params.tagSz - 1) / params.tagSz) map { j =>
+        io.f1_ghist(min((j + 1) * params.tagSz, bhLen) - 1, j * params.tagSz)
+      }).reduce(_^_)
+    } else {
+      0.U(params.tagSz.W)
+    }
+    val folded_phist = if (phLen > 0) {
+      ((0 until phLen) map { j => {
+        val ph = io.f1_phist(j * indirectHistoryWidth + params.tagSz - 1, j * indirectHistoryWidth)
+        Cat(if (j > 0) ph(j - 1, 0) else 0.U, ph(indirectHistoryWidth - 1, j)) // rotate right shift
+        // ph // uncomment for verification with no shift
+      }}).reduce(_^_)
+    } else {
+      0.U(params.tagSz.W)
+    }
+    // assign F1 stage inputs for prediction
+    t.io.s1_valid := s1_valid
+    t.io.s1_tag   := fetchIdx(s1_pc)(params.tagSz - 1, 0) ^ folded_ghist ^ folded_phist
+    // assign update information
+    val prv = update_meta.base_provider(update_idx).valid && i.U === update_meta.base_provider(update_idx).bits
+    val alc = update_meta.base_allocate(update_idx).valid && i.U === update_meta.base_allocate(update_idx).bits
+    t.io.update_valid_meta := update_valid && (prv || alc && s1_update.bits.cfi_mispredicted)
+    t.io.update_valid_tag  := update_valid && (prv || alc) && s1_update.bits.cfi_mispredicted
+    t.io.update_tag        := Mux(prv, update_meta.base_prv_tag, update_meta.base_alc_tag)
+    t.io.update_way        := Mux(prv, update_meta.base_prv_way(update_idx), update_meta.base_alc_way(update_idx))
+    t.io.update_cfi_idx    := update_idx
+    t.io.update_target     := s1_update.bits.target
+    t.io.update_plru       := update_meta.base_plru(update_idx)
+    t.io.update_meta       := Mux(s1_update.bits.cfi_mispredicted,
+      0.U, Mux(update_meta.base_alt_diff(update_idx), 1.U, update_meta.base_useful(update_idx))
+    )
+    // reset ubit periodically
+    val low  = log2Ceil(params.uBitPeriod)
+    val mid  = log2Ceil(params.uBitPeriod) + log2Ceil(nSets)
+    val high = log2Ceil(params.uBitPeriod) + log2Ceil(nSets) + log2Ceil(nWays)
+    val ubit_cnt       = RegInit(1.U((high).W))
+    val ubit_clear     = ubit_cnt(low - 1, 0) === 0.U
+    val ubit_clear_idx = ubit_cnt(mid - 1, low)
+    val ubit_clear_way = if (nWays > 1) ubit_cnt(high - 1, mid) else 0.U
+    when (!update_valid) { ubit_cnt := ubit_cnt + 1.U }
+    when (!update_valid && ubit_clear) {
+      t.io.update_valid_meta := true.B
+      t.io.update_meta       := 0.U
+      t.io.update_tag        := ubit_clear_idx
+      t.io.update_way        := ubit_clear_way
+    }
+    // the module
+    t
+  }
+  val s2_base_resp = WireInit(0.U.asTypeOf(s2_resp))
+  val random_allocate = RegInit(0.U(log2Ceil(params.baseTableInfo.length + 1).W))
+  when (random_allocate === (params.baseTableInfo.length - 1).U) {
+    random_allocate := 0.U
+  }.otherwise {
+    random_allocate := random_allocate + 1.U
+  }
+  for (w <- 0 until bankWidth) {
+    // find base_provider and altpred
+    var altpred = io.resp_in(0).f2(w).predicted_pc.bits
+    val final_altpred = WireInit(io.resp_in(0).f2(w).predicted_pc.bits)
+    for (i <- 0 until baseP.length) {
+      when (baseP(i).io.s2_resp(w).valid) {
+        s2_base_resp(w)                := baseP(i).io.s2_resp(w)
+        final_altpred                  := altpred
+        s2_meta.base_provider(w).valid := true.B
+        s2_meta.base_provider(w).bits  := i.U
+        s2_meta.base_useful(w)         := baseP(i).io.s2_meta(w)
+        s2_meta.base_plru(w)           := baseP(i).io.s2_plru(w)
+        s2_meta.base_prv_tag           := baseP(i).io.s2_tag
+        s2_meta.base_prv_way(w)        := baseP(i).io.s2_way(w)
+      }
+      altpred = Mux(baseP(i).io.s2_resp(w).valid, baseP(i).io.s2_resp(w).bits, altpred)
+    }
+    // first randomly set the base_allocate (in case all entries are useful)
+    for (i <- 0 until baseP.length) {
+      when (i.U === random_allocate) {
+        s2_meta.base_allocate(w).valid := true.B
+        s2_meta.base_allocate(w).bits  := i.U
+        s2_meta.base_alc_tag           := baseP(i).io.s2_tag
+        s2_meta.base_alc_way(w)        := baseP(i).io.s2_way(w)
+      }
+    }
+    // find not useful entry for allocation
+    for (i <- baseP.length - 1 to 0 by -1) {
+      when ((!s2_meta.base_provider(w).valid || i.U > s2_meta.base_provider(w).bits) && !baseP(i).io.s2_meta(w)) {
+        s2_meta.base_allocate(w).valid := true.B
+        s2_meta.base_allocate(w).bits  := i.U
+        s2_meta.base_alc_tag           := baseP(i).io.s2_tag
+        s2_meta.base_alc_way(w)        := baseP(i).io.s2_way(w)
+      }
+    }
+    s2_meta.base_alt_diff(w) := s2_base_resp(w).bits =/= final_altpred
+    s2_meta.base_hashed(w) := foldMetaResp(s2_base_resp(w).bits)
+  }
+
+  // load value correlation predictor
+  val lvcP = params.lvcTableInfo.zipWithIndex.map { case ((nSets, nWays), i) =>
+    val t = Module(new IndirectBranchPredictionTable(nSets, nWays, params.tagSz, params.lvcCntBits))
+    require(params.tagSz <= loadHistoryWidth)
+    // assign F1 stage inputs for prediction
+    t.io.s1_valid := s1_valid
+    t.io.s1_tag   := fetchIdx(s1_pc)(params.tagSz - 1, 0) ^
+      io.f1_lphist(i * loadHistoryWidth + params.tagSz - 1, i * loadHistoryWidth) ^
+      io.f1_lvhist(i * loadHistoryWidth + params.tagSz - 1, i * loadHistoryWidth)
+    // assign update information
+    val old_cnt = update_meta.lvc_cnt(update_idx)
+    val new_cnt = WireInit(old_cnt)
+    when (update_meta.lvc_hashed(update_idx) === update_folded_target) {
+      val saturated = (1.U << params.lvcCntBits) - 1.U
+      new_cnt := Mux(old_cnt === saturated, saturated, old_cnt + 1.U)
+    }.otherwise {
+      new_cnt := Mux(old_cnt === 0.U, 0.U, old_cnt - 1.U)
+    }
+    val prv = update_meta.lvc_provider(update_idx) === i.U
+    t.io.update_valid_meta := update_valid && prv
+    t.io.update_valid_tag  := update_valid && prv && old_cnt === 0.U
+    t.io.update_tag        := update_meta.lvc_prv_tag
+    t.io.update_way        := update_meta.lvc_prv_way(update_idx)
+    t.io.update_cfi_idx    := update_idx
+    t.io.update_target     := s1_update.bits.target
+    t.io.update_meta       := new_cnt
+    t.io.update_plru       := update_meta.lvc_plru(update_idx)
+    // the module
+    t
+  }
+  val s2_lvhist = VecInit((0 until params.lvcTableInfo.length) map {
+    i => RegNext(io.f1_lvhist((i + 1) * loadHistoryWidth - 1, i * loadHistoryWidth))
+  })
+
+  // lvc predictor response
+  val s2_lvc_resp = WireInit(0.U.asTypeOf(s2_resp))
+  val random_sel = RegInit(0.U(log2Ceil(params.lvcTableInfo.length).W))
+  when (random_sel === (params.lvcTableInfo.length - 1).U) {
+    random_sel := 0.U
+  }.otherwise {
+    random_sel := random_sel + 1.U
+  }
+  for (w <- 0 until bankWidth) {
+    // lvcP with no tag cannot recognize JALR and requires validation of base predictor response
+    when (s2_base_resp(w).valid) {
+      // select randomly in case of no saturated entry
+      for (i <- params.lvcTableInfo.length - 1 to 0 by -1) {
+        when (i.U === random_sel) {
+          s2_meta.lvc_provider(w) := i.U
+          s2_meta.lvc_hashed  (w) := foldMetaResp(lvcP(i).io.s2_resp(w).bits)
+          s2_meta.lvc_cnt     (w) := lvcP(i).io.s2_meta(w)
+          s2_meta.lvc_plru    (w) := lvcP(i).io.s2_plru(w)
+          s2_meta.lvc_prv_tag     := lvcP(i).io.s2_tag
+          s2_meta.lvc_prv_way (w) := lvcP(i).io.s2_way(w)
+        }
+      }
+      // find the latest saturated entry
+      for (i <- params.lvcTableInfo.length - 1 to 0 by -1) {
+        when (s2_lvhist(i) =/= 0.U && lvcP(i).io.s2_meta(w)(params.lvcCntBits - 1) === 1.U) {
+          s2_meta.lvc_provider(w) := i.U
+          s2_meta.lvc_hashed  (w) := foldMetaResp(lvcP(i).io.s2_resp(w).bits)
+          s2_meta.lvc_cnt     (w) := lvcP(i).io.s2_meta(w)
+          s2_meta.lvc_plru    (w) := lvcP(i).io.s2_plru(w)
+          s2_meta.lvc_prv_tag     := lvcP(i).io.s2_tag
+          s2_meta.lvc_prv_way (w) := lvcP(i).io.s2_way(w)
+          s2_lvc_resp         (w) := lvcP(i).io.s2_resp(w)
+        }
+      }
+    }
+  }
+
+  // assign responses
+  io.f3_meta := RegNext(s2_meta).asUInt
+  for (w <- 0 until bankWidth) {
+    s2_resp(w) := Mux(s2_lvcsel && s2_lvc_resp(w).valid, s2_lvc_resp(w), s2_base_resp(w))
+    when (RegNext(s2_resp(w).valid)) {
+      io.resp.f3(w).predicted_pc := RegNext(s2_resp(w))
+    }
+  }
+
+  override val metaSz = s2_meta.asUInt.getWidth
+  val mems = Nil
+}
diff --git a/src/main/scala/ifu/bpd/predictor.scala b/src/main/scala/ifu/bpd/predictor.scala
index f176c8df..6e3a7fb6 100644
--- a/src/main/scala/ifu/bpd/predictor.scala
+++ b/src/main/scala/ifu/bpd/predictor.scala
@@ -144,6 +144,9 @@ abstract class BranchPredictorBank(implicit p: Parameters) extends BoomModule()(
     // Local history not available until end of f1
     val f1_ghist = Input(UInt(globalHistoryLength.W))
     val f1_lhist = Input(UInt(localHistoryLength.W))
+    val f1_phist = Input(UInt((indirectHistoryLength * indirectHistoryWidth).W))
+    val f1_lphist = Input(UInt((loadHistoryLength * loadHistoryWidth).W))
+    val f1_lvhist = Input(UInt((loadHistoryLength * loadHistoryWidth).W))
 
     val resp_in = Input(Vec(nInputs, new BranchPredictionBankResponse))
     val resp = Output(new BranchPredictionBankResponse)
@@ -238,6 +241,9 @@ class BranchPredictor(implicit p: Parameters) extends BoomModule()(p)
 
     banked_predictors(0).io.f1_ghist := RegNext(io.f0_req.bits.ghist.histories(0))
     banked_predictors(0).io.f1_lhist := banked_lhist_providers(0).io.f1_lhist
+    banked_predictors(0).io.f1_phist := RegNext(io.f0_req.bits.ghist.path_history)
+    banked_predictors(0).io.f1_lphist := RegNext(io.f0_req.bits.ghist.load_pc_history)
+    banked_predictors(0).io.f1_lvhist := RegNext(io.f0_req.bits.ghist.load_val_history)
 
     banked_predictors(0).io.resp_in(0)           := (0.U).asTypeOf(new BranchPredictionBankResponse)
   } else {
@@ -249,6 +255,13 @@ class BranchPredictor(implicit p: Parameters) extends BoomModule()(p)
     banked_predictors(0).io.f1_lhist := banked_lhist_providers(0).io.f1_lhist
     banked_predictors(1).io.f1_lhist := banked_lhist_providers(1).io.f1_lhist
 
+    banked_predictors(0).io.f1_phist  := RegNext(io.f0_req.bits.ghist.path_history)
+    banked_predictors(1).io.f1_phist  := RegNext(io.f0_req.bits.ghist.path_history)
+    banked_predictors(0).io.f1_lphist := RegNext(io.f0_req.bits.ghist.load_pc_history)
+    banked_predictors(1).io.f1_lphist := RegNext(io.f0_req.bits.ghist.load_pc_history)
+    banked_predictors(0).io.f1_lvhist := RegNext(io.f0_req.bits.ghist.load_val_history)
+    banked_predictors(1).io.f1_lvhist := RegNext(io.f0_req.bits.ghist.load_val_history)
+
     when (bank(io.f0_req.bits.pc) === 0.U) {
       banked_lhist_providers(0).io.f0_valid := io.f0_req.valid
       banked_lhist_providers(0).io.f0_pc    := bankAlign(io.f0_req.bits.pc)
diff --git a/src/main/scala/ifu/fetch-target-queue.scala b/src/main/scala/ifu/fetch-target-queue.scala
index 5af74617..b452e6e7 100644
--- a/src/main/scala/ifu/fetch-target-queue.scala
+++ b/src/main/scala/ifu/fetch-target-queue.scala
@@ -155,6 +155,7 @@ class FetchTargetQueue(implicit p: Parameters) extends BoomModule
   val prev_ghist = RegInit((0.U).asTypeOf(new GlobalHistory))
   val prev_entry = RegInit((0.U).asTypeOf(new FTQBundle))
   val prev_pc    = RegInit(0.U(vaddrBitsExtended.W))
+  val prev_target = RegInit(0.U(vaddrBitsExtended.W))
   when (do_enq) {
 
     pcs(enq_ptr)           := io.enq.bits.pc
@@ -185,7 +186,9 @@ class FetchTargetQueue(implicit p: Parameters) extends BoomModule
         prev_entry.cfi_idx.valid,
         prev_pc,
         prev_entry.cfi_is_call,
-        prev_entry.cfi_is_ret
+        prev_entry.cfi_is_ret,
+        prev_entry.cfi_type === CFI_JALR && !prev_entry.cfi_is_ret,
+        prev_target
       )
     )
 
@@ -195,6 +198,7 @@ class FetchTargetQueue(implicit p: Parameters) extends BoomModule
     ram(enq_ptr) := new_entry
 
     prev_pc    := io.enq.bits.pc
+    prev_target := io.enq.bits.next_pc
     prev_entry := new_entry
     prev_ghist := new_ghist
 
@@ -294,6 +298,11 @@ class FetchTargetQueue(implicit p: Parameters) extends BoomModule
     io.bpdupdate.bits.target     := bpd_target
     io.bpdupdate.bits.cfi_is_br  := bpd_entry.br_mask(cfi_idx)
     io.bpdupdate.bits.cfi_is_jal := bpd_entry.cfi_type === CFI_JAL || bpd_entry.cfi_type === CFI_JALR
+    io.bpdupdate.bits.cfi_is_jalr := bpd_entry.cfi_type === CFI_JALR && !bpd_entry.cfi_is_ret &&
+    // The unequality condition is redundant because it is true when do_commit_update/is_commit_update
+    //   is true, but it results in different behavior if not added. Even just asserting the condition
+    //   can avoid this issue, possibly a Chisel bug.
+      RegNext(WrapInc(bpd_idx, num_entries) =/= enq_ptr)
     io.bpdupdate.bits.ghist      := bpd_ghist
     io.bpdupdate.bits.lhist      := bpd_lhist
     io.bpdupdate.bits.meta       := bpd_meta
@@ -310,6 +319,7 @@ class FetchTargetQueue(implicit p: Parameters) extends BoomModule
   val redirect_idx = io.redirect.bits
   val redirect_entry = ram(redirect_idx)
   val redirect_new_entry = WireInit(redirect_entry)
+  val redirect_new_target = WireInit(0.U(vaddrBitsExtended.W))
 
   when (io.redirect.valid) {
     enq_ptr    := WrapInc(io.redirect.bits, num_entries)
@@ -323,7 +333,11 @@ class FetchTargetQueue(implicit p: Parameters) extends BoomModule
       redirect_new_entry.cfi_taken        := io.brupdate.b2.taken
       redirect_new_entry.cfi_is_call      := redirect_entry.cfi_is_call && redirect_entry.cfi_idx.bits === new_cfi_idx
       redirect_new_entry.cfi_is_ret       := redirect_entry.cfi_is_ret  && redirect_entry.cfi_idx.bits === new_cfi_idx
+      when (redirect_entry.cfi_idx.bits =/= new_cfi_idx) {
+        redirect_new_entry.cfi_type := 0.U
+      }
     }
+    redirect_new_target := io.brupdate.b2.jalr_target
 
     ras_update     := true.B
     ras_update_pc  := redirect_entry.ras_top
@@ -333,6 +347,7 @@ class FetchTargetQueue(implicit p: Parameters) extends BoomModule
     prev_entry := RegNext(redirect_new_entry)
     prev_ghist := bpd_ghist
     prev_pc    := bpd_pc
+    prev_target := RegNext(redirect_new_target)
 
     ram(RegNext(io.redirect.bits)) := RegNext(redirect_new_entry)
   }
diff --git a/src/main/scala/ifu/frontend.scala b/src/main/scala/ifu/frontend.scala
index 17bf3b2e..e6c89007 100644
--- a/src/main/scala/ifu/frontend.scala
+++ b/src/main/scala/ifu/frontend.scala
@@ -57,6 +57,11 @@ class GlobalHistory(implicit p: Parameters) extends BoomBundle()(p)
 
   val ras_idx = UInt(log2Ceil(nRasEntries).W)
 
+  val path_history = UInt((indirectHistoryLength * indirectHistoryWidth).W)
+
+  val load_pc_history = UInt((loadHistoryLength * loadHistoryWidth).W)
+  val load_val_history = UInt((loadHistoryLength * loadHistoryWidth).W)
+
   def histories(bank: Int) = {
     if (nBanks == 1) {
       old_history
@@ -75,14 +80,16 @@ class GlobalHistory(implicit p: Parameters) extends BoomBundle()(p)
   def ===(other: GlobalHistory): Bool = {
     ((old_history === other.old_history) &&
      (new_saw_branch_not_taken === other.new_saw_branch_not_taken) &&
-     (new_saw_branch_taken === other.new_saw_branch_taken)
+     (new_saw_branch_taken === other.new_saw_branch_taken) &&
+     (path_history === other.path_history)
     )
   }
   def =/=(other: GlobalHistory): Bool = !(this === other)
 
   def update(branches: UInt, cfi_taken: Bool, cfi_is_br: Bool, cfi_idx: UInt,
     cfi_valid: Bool, addr: UInt,
-    cfi_is_call: Bool, cfi_is_ret: Bool): GlobalHistory = {
+    cfi_is_call: Bool, cfi_is_ret: Bool,
+    cfi_is_jalr: Bool = false.B, cfi_target: UInt = 0.U): GlobalHistory = {
     val cfi_idx_fixed = cfi_idx(log2Ceil(fetchWidth)-1,0)
     val cfi_idx_oh = UIntToOH(cfi_idx_fixed)
     val new_history = Wire(new GlobalHistory)
@@ -123,6 +130,10 @@ class GlobalHistory(implicit p: Parameters) extends BoomBundle()(p)
     }
     new_history.ras_idx := Mux(cfi_valid && cfi_is_call, WrapInc(ras_idx, nRasEntries),
                            Mux(cfi_valid && cfi_is_ret , WrapDec(ras_idx, nRasEntries), ras_idx))
+    new_history.path_history := Mux(
+      cfi_valid && cfi_is_jalr,
+      path_history << indirectHistoryWidth | (cfi_target & (1.U << indirectHistoryWidth) - 1.U),
+      path_history)
     new_history
   }
 
@@ -285,7 +296,15 @@ class BoomFrontendIO(implicit p: Parameters) extends BoomBundle
 
   val flush_icache = Output(Bool())
 
+  val load_pc_history = Output(UInt((loadHistoryLength * loadHistoryWidth).W))
+  val load_val_history = Output(UInt((loadHistoryLength * loadHistoryWidth).W))
+
   val perf = Input(new FrontendPerfEvents)
+
+  val tperf = Input(new Bundle {
+    val tlb_miss = Vec(ftqSz, Bool())
+    val cache_miss = Vec(ftqSz, Bool())
+  })
 }
 
 /**
@@ -799,7 +818,9 @@ class BoomFrontendModule(outer: BoomFrontend) extends LazyModuleImp(outer)
     f3_fetch_bundle.cfi_idx.valid,
     f3_fetch_bundle.pc,
     f3_fetch_bundle.cfi_is_call,
-    f3_fetch_bundle.cfi_is_ret
+    f3_fetch_bundle.cfi_is_ret,
+    f3_fetch_bundle.cfi_type === CFI_JALR && !f3_fetch_bundle.cfi_is_ret,
+    f3_fetch_bundle.next_pc
   )
 
 
@@ -982,9 +1003,52 @@ class BoomFrontendModule(outer: BoomFrontend) extends LazyModuleImp(outer)
     ftq.io.redirect.bits  := io.cpu.redirect_ftq_idx
   }
 
+  s0_ghist.load_pc_history := io.cpu.load_pc_history
+  s0_ghist.load_val_history := io.cpu.load_val_history
+
   ftq.io.debug_ftq_idx := io.cpu.debug_ftq_idx
   io.cpu.debug_fetch_pc := ftq.io.debug_fetch_pc
 
+  // jxy: performance monitor
+  val tlbVadd = RegInit(0.U(vaddrBitsExtended.W))
+  val cachePadd = RegInit(0.U(paddrBits.W))
+  val f3_tlbMiss = withReset(reset.asBool || f3_clear) { Module(new Queue(Bool(), 1, pipe=true, flow=false)) }
+  val f3_cacheMiss = withReset(reset.asBool || f3_clear) { Module(new Queue(Bool(), 1, pipe=true, flow=false)) }
+  val f4_tlbMiss = withReset(reset.asBool || f4_clear) { Module(new Queue(Bool(), 1, pipe=true, flow=false)) }
+  val f4_cacheMiss = withReset(reset.asBool || f4_clear) { Module(new Queue(Bool(), 1, pipe=true, flow=false)) }
+  val ftq_enq_ptr = RegInit(1.U(log2Ceil(ftqSz).W))
+  val ftq_tlbMiss = Reg(Vec(ftqSz, Bool()))
+  val ftq_cacheMiss = Reg(Vec(ftqSz, Bool()))
+  f3_tlbMiss  .io.enq.valid := f3.io.enq.valid
+  f3_cacheMiss.io.enq.valid := f3.io.enq.valid
+  f3_tlbMiss  .io.enq.bits  := (s2_vpc >> pgIdxBits)    ===  tlbVadd
+  f3_cacheMiss.io.enq.bits  := (s2_ppc >> blockOffBits) === (cachePadd >> blockOffBits)
+  f3_tlbMiss  .io.deq.ready := f3.io.deq.ready
+  f3_cacheMiss.io.deq.ready := f3.io.deq.ready
+  f4_tlbMiss  .io.enq.valid := f4.io.enq.valid
+  f4_cacheMiss.io.enq.valid := f4.io.enq.valid
+  f4_tlbMiss  .io.enq.bits  := f3_tlbMiss  .io.deq.bits
+  f4_cacheMiss.io.enq.bits  := f3_cacheMiss.io.deq.bits
+  f4_tlbMiss  .io.deq.ready := f4.io.deq.ready
+  f4_cacheMiss.io.deq.ready := f4.io.deq.ready
+  when (f3_tlbMiss  .io.enq.valid && f3_tlbMiss  .io.enq.bits) { tlbVadd   := 0.U }
+  when (f3_cacheMiss.io.enq.valid && f3_cacheMiss.io.enq.bits) { cachePadd := 0.U }
+  when (tlb   .io.perf.miss)    { tlbVadd   := tlb   .io.perf.vadd }
+  when (icache.io.perf.acquire) { cachePadd := icache.io.perf.padd }
+  when (ftq.io.deq.valid) {
+    ftq_tlbMiss  (ftq.io.deq.bits) := false.B
+    ftq_cacheMiss(ftq.io.deq.bits) := false.B
+  }
+  when (ftq.io.enq.fire) {
+    ftq_tlbMiss  (ftq_enq_ptr) := f4_tlbMiss  .io.deq.bits
+    ftq_cacheMiss(ftq_enq_ptr) := f4_cacheMiss.io.deq.bits
+    ftq_enq_ptr := WrapInc(ftq_enq_ptr, ftqSz)
+  }
+  when (ftq.io.redirect.valid) {
+    ftq_enq_ptr := WrapInc(ftq.io.redirect.bits, ftqSz)
+  }
+  io.cpu.tperf.tlb_miss   := ftq_tlbMiss
+  io.cpu.tperf.cache_miss := ftq_cacheMiss
 
   override def toString: String =
     (BoomCoreStringPrefix("====Overall Frontend Params====") + "\n"
diff --git a/src/main/scala/ifu/icache.scala b/src/main/scala/ifu/icache.scala
index 916a688f..e7181fc1 100644
--- a/src/main/scala/ifu/icache.scala
+++ b/src/main/scala/ifu/icache.scala
@@ -82,6 +82,7 @@ class ICacheBundle(val outer: ICache) extends BoomBundle()(outer.p)
 
   val perf = Output(new Bundle {
     val acquire = Bool()
+    val padd = UInt(paddrBits.W)
   })
 }
 
@@ -334,6 +335,7 @@ class ICacheModule(outer: ICache) extends LazyModuleImp(outer)
   tl_out.e.valid := false.B
 
   io.perf.acquire := tl_out.a.fire
+  io.perf.padd := refill_paddr
 
   when (!refill_valid) { invalidated := false.B }
   when (refill_fire) { refill_valid := true.B }
diff --git a/src/main/scala/lsu/dcache.scala b/src/main/scala/lsu/dcache.scala
index 9f326f50..64fae1b0 100644
--- a/src/main/scala/lsu/dcache.scala
+++ b/src/main/scala/lsu/dcache.scala
@@ -820,6 +820,7 @@ class BoomNonBlockingDCacheModule(outer: BoomNonBlockingDCache) extends LazyModu
 
   io.lsu.perf.release := edge.done(tl_out.c)
   io.lsu.perf.acquire := edge.done(tl_out.a)
+  io.lsu.perf.padd := tl_out.a.bits.address
 
   // load data gen
   val s2_data_word_prebypass = widthMap(w => s2_data_muxed(w) >> Cat(s2_word_idx(w), 0.U(log2Ceil(coreDataBits).W)))
diff --git a/src/main/scala/lsu/lsu.scala b/src/main/scala/lsu/lsu.scala
index 1eb8ba24..ef57c55d 100644
--- a/src/main/scala/lsu/lsu.scala
+++ b/src/main/scala/lsu/lsu.scala
@@ -102,6 +102,7 @@ class LSUDMemIO(implicit p: Parameters, edge: TLEdgeOut) extends BoomBundle()(p)
   val perf = Input(new Bundle {
     val acquire = Bool()
     val release = Bool()
+    val padd = UInt(paddrBits.W)
   })
 
 }
@@ -153,6 +154,11 @@ class LSUCoreIO(implicit p: Parameters) extends BoomBundle()(p)
     val release = Bool()
     val tlbMiss = Bool()
   })
+
+  val tperf = Output(Vec(memWidth, new Bundle {
+    val tlb_miss = Valid(UInt(robAddrSz.W))
+    val cache_miss = Valid(UInt(robAddrSz.W))
+  }))
 }
 
 class LSUIO(implicit p: Parameters, edge: TLEdgeOut) extends BoomBundle()(p)
@@ -1651,6 +1657,32 @@ class LSU(implicit p: Parameters, edge: TLEdgeOut) extends BoomModule()(p)
                     ~(st_exc_killed_mask.asUInt)
 
 
+  // jxy: performance monitor
+  val tlbVadd = RegInit((0.U(vaddrBitsExtended.W)))
+  val cachePadd = RegInit(0.U(coreMaxAddrBits.W))
+  for (w <- 0 until memWidth) {
+    io.core.tperf(w).tlb_miss.valid := false.B
+    io.core.tperf(w).cache_miss.valid := false.B
+    io.core.tperf(w).tlb_miss.bits := DontCare
+    io.core.tperf(w).cache_miss.bits := DontCare
+    when (exe_tlb_valid(w) && (exe_tlb_vaddr(w) >> pgIdxBits) === tlbVadd) {
+      io.core.tperf(w).tlb_miss.valid := true.B
+      io.core.tperf(w).tlb_miss.bits := exe_tlb_uop(w).rob_idx
+      tlbVadd := 0.U
+    }
+    val resp_addr = Mux(
+      io.dmem.resp(w).bits.uop.uses_ldq,
+      ldq(io.dmem.resp(w).bits.uop.ldq_idx).bits.addr.bits,
+      stq(io.dmem.resp(w).bits.uop.stq_idx).bits.addr.bits
+    )
+    when (io.dmem.resp(w).valid && (resp_addr >> blockOffBits) === (cachePadd >> blockOffBits)) {
+      io.core.tperf(w).cache_miss.valid := true.B
+      io.core.tperf(w).cache_miss.bits := io.dmem.resp(w).bits.uop.rob_idx
+      cachePadd := 0.U
+    }
+  }
+  when (dtlb.io.perf.miss)    { tlbVadd   := dtlb.io.perf.vadd }
+  when (io.dmem.perf.acquire) { cachePadd := io.dmem.perf.padd }
 }
 
 /**
diff --git a/src/main/scala/lsu/tlb.scala b/src/main/scala/lsu/tlb.scala
index a09018b4..311b0412 100644
--- a/src/main/scala/lsu/tlb.scala
+++ b/src/main/scala/lsu/tlb.scala
@@ -23,6 +23,11 @@ class NBDTLB(instruction: Boolean, lgMaxSize: Int, cfg: TLBConfig)(implicit edge
     val sfence = Input(Valid(new SFenceReq))
     val ptw = new TLBPTWIO
     val kill = Input(Bool())
+    // jxy: performance monitor
+    val perf = Output(new Bundle {
+      val miss = Output(Bool())
+      val vadd = Output(UInt(vaddrBitsExtended.W))
+    })
   })
   io.ptw := DontCare
   io.resp := DontCare
@@ -358,4 +363,7 @@ class NBDTLB(instruction: Boolean, lgMaxSize: Int, cfg: TLBConfig)(implicit edge
   }
 
 
+  // jxy: performance monitor
+  io.perf.miss := io.ptw.req.fire
+  io.perf.vadd := io.ptw.req.bits.bits.addr
 }
