diff --git a/src/main/scala/common/config-mixins.scala b/src/main/scala/common/config-mixins.scala
index 5de8ebf4..84798e05 100644
--- a/src/main/scala/common/config-mixins.scala
+++ b/src/main/scala/common/config-mixins.scala
@@ -156,7 +156,7 @@ class WithNMediumBooms(n: Int = 1, overrideIdOffset: Option[Int] = None) extends
               fpu = Some(freechips.rocketchip.tile.FPUParams(sfmaLatency=4, dfmaLatency=4, divSqrt=true))
             ),
             dcache = Some(
-              DCacheParams(rowBits = 64, nSets=64, nWays=4, nMSHRs=2, nTLBWays=8)
+              DCacheParams(rowBits = 64, nSets=64, nWays=4, nMSHRs=2, nTLBWays=16)
             ),
             icache = Some(
               ICacheParams(rowBits = 64, nSets=64, nWays=4, fetchBytes=2*4)
@@ -437,23 +437,25 @@ class WithNCS152DefaultBooms(n: Int = 1, overrideIdOffset: Option[Int] = None) e
 class WithTAGELBPD extends Config((site, here, up) => {
   case TilesLocated(InSubsystem) => up(TilesLocated(InSubsystem), site) map {
     case tp: BoomTileAttachParams => tp.copy(tileParams = tp.tileParams.copy(core = tp.tileParams.core.copy(
-      bpdMaxMetaLength = 120,
+      bpdMaxMetaLength = 140,
       globalHistoryLength = 64,
       localHistoryLength = 1,
       localHistoryNSets = 0,
       branchPredictor = ((resp_in: BranchPredictionBankResponse, p: Parameters) => {
         val loop = Module(new LoopBranchPredictorBank()(p))
         val tage = Module(new TageBranchPredictorBank()(p))
+        val ibp = Module(new IndirectBranchPredictorBank()(p))
         val btb = Module(new BTBBranchPredictorBank()(p))
         val bim = Module(new BIMBranchPredictorBank()(p))
         val ubtb = Module(new FAMicroBTBBranchPredictorBank()(p))
-        val preds = Seq(loop, tage, btb, ubtb, bim)
+        val preds = Seq(loop, tage, ibp, btb, ubtb, bim)
         preds.map(_.io := DontCare)
 
         ubtb.io.resp_in(0)  := resp_in
         bim.io.resp_in(0)   := ubtb.io.resp
         btb.io.resp_in(0)   := bim.io.resp
-        tage.io.resp_in(0)  := btb.io.resp
+        ibp.io.resp_in(0)   := btb.io.resp
+        tage.io.resp_in(0)  := ibp.io.resp
         loop.io.resp_in(0)  := tage.io.resp
 
         (preds, loop.io.resp)
diff --git a/src/main/scala/common/parameters.scala b/src/main/scala/common/parameters.scala
index 47da9ee5..d85443cf 100644
--- a/src/main/scala/common/parameters.scala
+++ b/src/main/scala/common/parameters.scala
@@ -65,6 +65,8 @@ case class BoomCoreParams(
   bpdMaxMetaLength: Int = 120,
   numRasEntries: Int = 32,
   enableRasTopRepair: Boolean = true,
+  indirectHistoryLength: Int = 4,
+  indirectHistoryWidth: Int = 16,
 
   /* more stuff */
   useCompressed: Boolean = true,
@@ -257,6 +259,9 @@ trait HasBoomCoreParameters extends freechips.rocketchip.tile.HasCoreParameters
 
   val useLHist = localHistoryNSets > 1 && localHistoryLength > 1
 
+  val indirectHistoryLength = boomParams.indirectHistoryLength
+  val indirectHistoryWidth = boomParams.indirectHistoryWidth
+
   //************************************
   // Extra Knobs and Features
   val enableCommitMapTable = boomParams.enableCommitMapTable
diff --git a/src/main/scala/common/tile.scala b/src/main/scala/common/tile.scala
index 05821c52..dd1c9032 100644
--- a/src/main/scala/common/tile.scala
+++ b/src/main/scala/common/tile.scala
@@ -6,7 +6,7 @@
 package boom.common
 
 import chisel3._
-import chisel3.util.{RRArbiter, Queue}
+import chisel3.util._
 
 import scala.collection.mutable.{ListBuffer}
 
@@ -243,6 +243,9 @@ class BoomTileModuleImp(outer: BoomTile) extends BaseTileModuleImp(outer){
   lsu.io.hellacache <> hellaCacheArb.io.mem
   outer.dcache.module.io.lsu <> lsu.io.dmem
 
+  // External performance monitor
+  outer.pmSourceNode.bundle <> core.io.pm
+
   // Generate a descriptive string
   val frontendStr = outer.frontend.module.toString
   val coreStr = core.toString
diff --git a/src/main/scala/exu/core.scala b/src/main/scala/exu/core.scala
index 0f8d39e3..82fb5be8 100644
--- a/src/main/scala/exu/core.scala
+++ b/src/main/scala/exu/core.scala
@@ -37,7 +37,7 @@ import org.chipsalliance.cde.config.Parameters
 import freechips.rocketchip.rocket.Instructions._
 import freechips.rocketchip.tile.{TraceBundle}
 import freechips.rocketchip.rocket.{Causes, PRV, TracedInstruction}
-import freechips.rocketchip.util.{Str, UIntIsOneOf, CoreMonitorBundle}
+import freechips.rocketchip.util.{Str, UIntIsOneOf, CoreMonitorBundle, PmSignals}
 import freechips.rocketchip.devices.tilelink.{PLICConsts, CLINTConsts}
 
 import boom.common._
@@ -61,6 +61,7 @@ class BoomCore()(implicit p: Parameters) extends BoomModule
     val ptw_tlb = new freechips.rocketchip.rocket.TLBPTWIO()
     val trace = Output(new TraceBundle)
     val fcsr_rm = UInt(freechips.rocketchip.tile.FPConstants.RM_SZ.W)
+    val pm = Output(new PmSignals)
   })
 
   io.ptw_tlb := DontCare
@@ -448,7 +449,9 @@ class BoomCore()(implicit p: Parameters) extends BoomModule
       true.B,
       io.ifu.get_pc(1).pc,
       ftq_entry.cfi_is_call && ftq_entry.cfi_idx.bits === cfi_idx,
-      ftq_entry.cfi_is_ret  && ftq_entry.cfi_idx.bits === cfi_idx)
+      ftq_entry.cfi_is_ret  && ftq_entry.cfi_idx.bits === cfi_idx,
+      ftq_entry.cfi_type === CFI_JALR && !ftq_entry.cfi_is_ret && ftq_entry.cfi_idx.bits === cfi_idx,
+      brupdate.b2.jalr_target)
 
 
     io.ifu.redirect_ghist   := Mux(
@@ -1492,4 +1495,227 @@ class BoomCore()(implicit p: Parameters) extends BoomModule
   } else {
     io.ifu.debug_ftq_idx := DontCare
   }
+
+  // jxy: performance monitor
+  val maskWidth = 4
+  val pidWidth = 11
+  val evtWidth = 3
+  val stageWidth = 2
+  val cntWidth = 4
+  val pipLatency = 14.U
+  val level = Wire(UInt(2.W)) // global
+  val pid   = RegInit(0.U(pidWidth.W))
+  val pc    = Wire(Vec(coreWidth, UInt(vaddrBits.W))) // per instruction
+  val addr  = Wire(Vec(coreWidth, UInt(vaddrBits.W)))
+  val stage = Wire(Vec(coreWidth, UInt(stageWidth.W)))
+  val evt   = RegInit(VecInit(Seq.fill(coreWidth)(0.U(evtWidth.W))))
+  val evm   = Wire(Vec(coreWidth, Vec(maskWidth, Bool())))
+  for (w <- 0 until coreWidth) {
+    io.pm.tag(w)   := RegNext(Cat(pid, level, evt(w)))
+    io.pm.info(w)  := RegNext(Cat(addr(w), stage(w), pc(w)))
+    io.pm.event(w) := RegNext(evm(w).asUInt)
+  }
+  // Additional info in ROB
+  val next_pc_reg = RegInit(0.U(vaddrBits.W))
+  val last_pc_reg = RegInit(0.U(vaddrBits.W))
+  val last_cnt_reg = RegInit(0.U(cntWidth.W))
+  val last_evm_reg = RegInit(VecInit(Seq.fill(maskWidth)(false.B)))
+  val next_pc = Wire(UInt(vaddrBits.W))
+  val last_pc = Wire(UInt(vaddrBits.W))
+  val last_cnt = Wire(UInt(cntWidth.W))
+  val last_evm = Wire(Vec(maskWidth, Bool()))
+  val next_evm = Wire(Vec(maskWidth, Bool()))
+  val next_stage = Wire(UInt(stageWidth.W))
+  next_pc_reg := next_pc
+  last_pc_reg := last_pc
+  last_cnt_reg := last_cnt
+  last_evm_reg := last_evm
+  next_pc := next_pc_reg
+  last_pc := last_pc_reg
+  last_cnt := Mux(last_cnt_reg =/= 0.U && rob.io.ready, last_cnt_reg - 1.U, last_cnt_reg)
+  last_evm := last_evm_reg
+  next_stage := 0.U
+  next_evm := last_evm
+  for (w <- 0 until coreWidth) {
+    def GetRowIdx(rob_idx: UInt): UInt = {
+      if (coreWidth == 1) return rob_idx
+      else return rob_idx >> log2Ceil(coreWidth).U
+    }
+    def GetBankIdx(rob_idx: UInt): UInt = {
+      if (coreWidth == 1) return 0.U
+      else return rob_idx(log2Ceil(coreWidth)-1, 0).asUInt
+    }
+    def MatchBank(bank_idx: UInt): Bool = (bank_idx === w.U)
+    val rob_stage = RegInit(VecInit(Seq.fill(numRobRows)(0.U(stageWidth.W))))
+    val rob_pid = RegInit(VecInit(Seq.fill(numRobRows)(0.U(pidWidth.W))))
+    val rob_misp = RegInit(VecInit(Seq.fill(numRobRows)(false.B)))
+    val rob_taken = RegInit(VecInit(Seq.fill(numRobRows)(false.B)))
+    val rob_addr = RegInit(VecInit(Seq.fill(numRobRows)(0.U(vaddrBits.W))))
+    val rob_dtm = RegInit(VecInit(Seq.fill(numRobRows)(false.B)))
+    val rob_dcm = RegInit(VecInit(Seq.fill(numRobRows)(false.B)))
+    val rob_cnt = RegInit(VecInit(Seq.fill(numRobRows)(0.U(cntWidth.W))))
+    val com_pc = Wire(UInt(vaddrBits.W))
+    val com_cnt = Wire(UInt(cntWidth.W))
+    val com_evm = Wire(Vec(maskWidth, Bool()))
+    val com_addr = Wire(UInt(vaddrBits.W))
+    val com_evt = Wire(UInt(evtWidth.W))
+    // Count for redirect scope
+    for (i <- 0 until numRobRows) {
+      when (rob_cnt(i) =/= 0.U && rob.io.ready) {
+        rob_cnt(i) := rob_cnt(i) - 1.U
+      }
+    }
+    // Dispatching
+    for (i <- 0 until coreWidth) {
+      when(dis_valids(i) && MatchBank(GetBankIdx(dis_uops(i).rob_idx))) {
+        rob_stage(GetRowIdx(dis_uops(i).rob_idx)) := 1.U
+        rob_cnt(GetRowIdx(dis_uops(i).rob_idx)) := 0.U
+        rob_addr(GetRowIdx(dis_uops(i).rob_idx)) := 0.U
+        rob_misp(GetRowIdx(dis_uops(i).rob_idx)) := false.B
+        rob_dtm(GetRowIdx(dis_uops(i).rob_idx)) := false.B
+        rob_dcm(GetRowIdx(dis_uops(i).rob_idx)) := false.B
+      }
+    }
+    // Issuing
+    for (i <- 0 until exe_units.length) {
+      when (iss_valids(i) && MatchBank(GetBankIdx(iss_uops(i).rob_idx))) {
+        rob_stage(GetRowIdx(iss_uops(i).rob_idx)) := 2.U
+      }
+    }
+    // Record PID from HINT argument
+    for (i <- 0 until rob.numWakeupPorts) {
+      val wb_resp = rob.io.wb_resps(i)
+      val wb_uop = wb_resp.bits.uop
+      when (wb_resp.valid && MatchBank(GetBankIdx(wb_uop.rob_idx))) {
+        rob_stage(GetRowIdx(wb_uop.rob_idx)) := 3.U
+        rob_pid(GetRowIdx(wb_uop.rob_idx)) := wb_resp.bits.data.asUInt
+      }
+    }
+    // Record branches and PCs
+    for (b <- brinfos) {
+      when (b.valid && MatchBank(GetBankIdx(b.uop.rob_idx))) {
+        rob_cnt(GetRowIdx(b.uop.rob_idx)) := Mux(b.mispredict, pipLatency, 0.U)
+        rob_misp(GetRowIdx(b.uop.rob_idx)) := b.mispredict
+        rob_taken(GetRowIdx(b.uop.rob_idx)) := b.taken
+        rob_addr(GetRowIdx(b.uop.rob_idx)) := b.jalr_target
+      }
+    }
+    // Record memory instruction addresses
+    for (i <- 0 until memWidth) {
+      when (io.lsu.exe(i).req.valid) {
+        when (MatchBank(GetBankIdx(io.lsu.exe(i).req.bits.uop.rob_idx))) {
+          rob_addr(GetRowIdx(io.lsu.exe(i).req.bits.uop.rob_idx)) := io.lsu.exe(i).req.bits.addr
+        }
+      }
+    }
+    // Record DTLB and data cache misses
+    for (i <- 0 until memWidth) {
+      when (io.lsu.tperf(i).tlb_miss.valid && MatchBank(GetBankIdx(io.lsu.tperf(i).tlb_miss.bits))) {
+        rob_dtm(GetRowIdx(io.lsu.tperf(i).tlb_miss.bits)) := true.B
+      }
+      when (io.lsu.tperf(i).cache_miss.valid && MatchBank(GetBankIdx(io.lsu.tperf(i).cache_miss.bits))) {
+        rob_dcm(GetRowIdx(io.lsu.tperf(i).cache_miss.bits)) := true.B
+      }
+    }
+    // Record ITLB and instruction cache miss
+    val itm = Wire(Bool())
+    val icm = Wire(Bool())
+    itm := io.ifu.tperf.tlb_miss(rob.io.commit.uops(w).ftq_idx)
+    icm := io.ifu.tperf.cache_miss(rob.io.commit.uops(w).ftq_idx)
+    for (i <- 0 until w) {
+      when (rob.io.commit.uops(w).ftq_idx === rob.io.commit.uops(i).ftq_idx) {
+        itm := false.B
+        icm := false.B
+      }
+    }
+    // Committing
+    when (rob.io.commit.valids(w)) {
+      rob_stage(GetRowIdx(rob.io.commit.uops(w).rob_idx)) := 0.U
+      rob_misp(GetRowIdx(rob.io.commit.uops(w).rob_idx)) := false.B
+      rob_dtm(GetRowIdx(rob.io.commit.uops(w).rob_idx)) := false.B
+      rob_dcm(GetRowIdx(rob.io.commit.uops(w).rob_idx)) := false.B
+    }
+    val uop = RegNext(rob.io.commit.uops(w))
+    when (RegNext(rob.io.commit.valids(w))) {
+      last_evm := RegNext(com_evm)
+      last_pc := com_pc
+      last_cnt := com_cnt
+      when (uop.is_br && rob_taken(GetRowIdx(uop.rob_idx))) {
+        next_pc := (com_pc.asSInt + ImmGen(uop.imm_packed, IS_B)).asUInt
+      }.elsewhen (uop.is_jal) {
+        next_pc := (com_pc.asSInt + ImmGen(uop.imm_packed, IS_J)).asUInt
+      }.elsewhen (uop.is_jalr) {
+        next_pc := rob_addr(GetRowIdx(uop.rob_idx))
+      }.otherwise {
+        next_pc := com_pc + Mux(uop.is_rvc, 2.U, 4.U)
+      }
+    }
+    when (RegNext(rob.io.flush.valid)) {
+      last_cnt := pipLatency
+      next_pc := io.ifu.redirect_pc
+    }
+    when (MatchBank(GetBankIdx(rob.io.rob_head_idx))) {
+      next_stage := Mux(rob.io.empty, 0.U, rob_stage(GetRowIdx(rob.io.rob_head_idx)))
+      next_evm := Mux(rob.io.empty, VecInit(Seq.fill(maskWidth)(false.B)), com_evm)
+      next_evm(0) := 1.B
+    }
+    // Get stage, event mask and PC
+    io.ifu.debug_ftq_idx(w) := rob.io.commit.uops(w).ftq_idx
+    com_pc := (AlignPCToBoundary(io.ifu.debug_fetch_pc(w), icBlockBytes)
+      + uop.pc_lob - Mux(uop.edge_inst, 2.U, 0.U))(vaddrBits-1,0)
+    com_cnt := rob_cnt(GetRowIdx(uop.rob_idx))
+    stage(w) := 0.U
+    com_evm(0) := 1.B
+    com_evm(1) := itm || rob_dtm(GetRowIdx(rob.io.commit.uops(w).rob_idx))
+    com_evm(2) := icm
+    com_evm(3) :=
+      rob_dcm(GetRowIdx(rob.io.commit.uops(w).rob_idx)) ||
+      rob_misp(GetRowIdx(rob.io.commit.uops(w).rob_idx))
+    com_addr := rob_addr(GetRowIdx(rob.io.commit.uops(w).rob_idx))
+    when (evt(w) === 0.U) {
+      when (last_cnt === 0.U) {
+        pc(w) := next_pc
+        evm(w) := next_evm
+        stage(w) := next_stage
+        addr(w) := 0.U
+      }.otherwise {
+        pc(w) := last_pc
+        evm(w) := last_evm
+        addr(w) := 0.U
+      }
+    }.otherwise {
+      pc(w) := com_pc
+      evm(w) := RegNext(com_evm)
+      addr(w) := RegNext(com_addr)
+    }
+    // Get PID
+    when (
+      rob.io.commit.valids(w) &&
+      rob.io.commit.uops(w).uopc === uopSLLI &&
+      rob.io.commit.uops(w).pdst === 0.U
+    ) {
+      pid := rob_pid(GetRowIdx(rob.io.commit.uops(w).rob_idx))
+    }
+    // Get event tag
+    when (rob.io.commit.uops(w).is_jalr && rob.io.commit.uops(w).ldst === 0.U && (
+      rob.io.commit.uops(w).lrs1 === 1.U || rob.io.commit.uops(w).lrs1 === 5.U
+    )) {
+      com_evt := 1.U // EVT::RET
+    }.elsewhen (rob.io.commit.uops(w).is_jalr) {
+      com_evt := 2.U // EVT:JALR
+    }.elsewhen (rob.io.commit.uops(w).is_br) {
+      com_evt := 3.U // EVT:BRANCH
+    }.elsewhen (rob.io.commit.uops(w).is_jal) {
+      com_evt := 4.U // EVT:JAL
+    }.otherwise {
+      com_evt := 5.U // EVT::OTHERS
+    }
+    when (!rob.io.commit.valids(w)) {
+      evt(w) := 0.U // EVT::NULL
+    }.otherwise {
+      evt(w) := com_evt
+    }
+  }
+  // Get privilege level
+  level := csr.io.status.prv
 }
diff --git a/src/main/scala/ifu/bpd/composer.scala b/src/main/scala/ifu/bpd/composer.scala
index 997c9b62..24e72f12 100644
--- a/src/main/scala/ifu/bpd/composer.scala
+++ b/src/main/scala/ifu/bpd/composer.scala
@@ -26,6 +26,7 @@ class ComposedBranchPredictorBank(implicit p: Parameters) extends BranchPredicto
     c.io.f0_mask   := io.f0_mask
     c.io.f1_ghist  := io.f1_ghist
     c.io.f1_lhist  := io.f1_lhist
+    c.io.f1_phist  := io.f1_phist
     c.io.f3_fire   := io.f3_fire
     if (c.metaSz > 0) {
       metas = (metas << c.metaSz) | c.io.f3_meta(c.metaSz-1,0)
diff --git a/src/main/scala/ifu/bpd/ibp.scala b/src/main/scala/ifu/bpd/ibp.scala
new file mode 100644
index 00000000..798332ed
--- /dev/null
+++ b/src/main/scala/ifu/bpd/ibp.scala
@@ -0,0 +1,169 @@
+package boom.ifu
+
+import chisel3._
+import chisel3.util._
+
+import org.chipsalliance.cde.config.{Field, Parameters}
+import freechips.rocketchip.diplomacy._
+import freechips.rocketchip.tilelink._
+
+import boom.common._
+import boom.util.{BoomCoreStringPrefix}
+
+import scala.math.min
+
+case class BoomIndirectParams(
+  bhLen: Int = 16,
+  phLen: Int = 4,
+  nSets: Int = 128,
+  nWays: Int = 2,
+  tagSz: Int = 16
+)
+
+
+class IndirectBranchPredictorBank(params: BoomIndirectParams = BoomIndirectParams())(implicit p: Parameters)
+  extends BranchPredictorBank()(p)
+{
+  override val nSets         = params.nSets
+  override val nWays         = params.nWays
+
+  class IBTBEntry extends Bundle {
+    val target = UInt(vaddrBitsExtended.W)
+  }
+  val ibtbEntrySz = vaddrBitsExtended
+
+  class IBTBMeta extends Bundle {
+    val tag = UInt(params.tagSz.W)
+  }
+  val ibtbMetaSz = params.tagSz
+
+  class IBTBPredictMeta extends Bundle {
+    val write_tag = UInt(params.tagSz.W)
+    val write_way = UInt(log2Ceil(nWays).W)
+  }
+
+  // reset
+  val doing_reset = RegInit(true.B)
+  val reset_idx   = RegInit(0.U(log2Ceil(nSets).W))
+  reset_idx := reset_idx + doing_reset
+  when (reset_idx === (nSets-1).U) { doing_reset := false.B }
+
+  // memory structure
+  val mems = Nil
+  val meta = Seq.fill(nWays) { SyncReadMem(nSets, Vec(bankWidth, UInt(ibtbMetaSz.W))) }
+  val ibtb = Seq.fill(nWays) { SyncReadMem(nSets, Vec(bankWidth, UInt(ibtbEntrySz.W))) }
+
+  // calculate folded path history
+  require(params.bhLen <= globalHistoryLength)
+  require(params.tagSz <= indirectHistoryWidth)
+  require(params.phLen <= indirectHistoryLength)
+  val folded_ghist = (0 until (params.bhLen + params.tagSz - 1) / params.tagSz) map {
+    i => io.f1_ghist(min((i + 1) * params.tagSz, params.bhLen) - 1, i * params.tagSz)
+  }
+  val folded_phist = (0 until params.phLen) map {
+    i => io.f1_phist(i * indirectHistoryWidth + params.tagSz - 1, i * indirectHistoryWidth)
+  }
+
+  // read/write ports and data
+  val s1_req_idx = Wire(UInt(log2Ceil(nSets).W))
+  val s1_req_tag = Wire(UInt(params.tagSz.W))
+  val s2_req_tag = RegNext(s1_req_tag)
+  val s2_req_ribtb = VecInit(ibtb.map {
+    b => VecInit(b.read(s1_req_idx, s1_valid).map(_.asTypeOf(new IBTBEntry)))
+  })
+  val s2_req_rmeta = VecInit(meta.map {
+    m => VecInit(m.read(s1_req_idx, s1_valid).map(_.asTypeOf(new IBTBMeta)))
+  })
+  s1_req_tag := fetchIdx(s1_pc) ^ folded_ghist.reduce(_^_) ^ folded_phist.reduce(_^_)
+  s1_req_idx := s1_req_tag(log2Ceil(nSets)-1, 0)
+
+  // tag matching and output results
+  val s2_resp = Wire(Vec(bankWidth, Valid(UInt(vaddrBitsExtended.W))))
+  val s2_hit_ohs = VecInit((0 until bankWidth) map { i =>
+    VecInit((0 until nWays) map { w =>
+      s2_req_rmeta(w)(i).tag === s2_req_tag
+    })
+  })
+  val s2_hits     = s2_hit_ohs.map { oh => oh.reduce(_||_) }
+  val s2_hit_ways = s2_hit_ohs.map { oh => PriorityEncoder(oh) }
+  for (w <- 0 until bankWidth) {
+    val entry_meta = s2_req_rmeta(s2_hit_ways(w))(w)
+    val entry_btb  = s2_req_ribtb(s2_hit_ways(w))(w)
+    s2_resp(w).valid := !doing_reset && s2_valid && s2_hits(w)
+    s2_resp(w).bits  := entry_btb.target
+
+    io.resp.f3(w) := io.resp_in(0).f3(w)
+    when (RegNext(s2_hits(w))) {
+      io.resp.f3(w).predicted_pc := RegNext(s2_resp(w))
+    }
+  }
+
+  // set metadata for write port
+  val s2_meta = Wire(new IBTBPredictMeta)
+  io.f3_meta := RegNext(s2_meta).asUInt
+  val alloc_way = if (nWays > 1) {
+    val r_metas = Cat(
+      VecInit(s2_req_rmeta.map { w => VecInit(w.map(_.tag)) }).asUInt,
+      s2_req_tag(params.tagSz-1,0)
+    )
+    val l = log2Ceil(nWays)
+    val nChunks = (r_metas.getWidth + l - 1) / l
+    val chunks = (0 until nChunks) map { i =>
+      r_metas(min((i+1)*l, r_metas.getWidth)-1, i*l)
+    }
+    chunks.reduce(_^_)
+  } else {
+    0.U
+  }
+  s2_meta.write_tag := RegNext(s1_req_tag)
+  s2_meta.write_way := Mux(s2_hits.reduce(_||_),
+    PriorityEncoder(s2_hit_ohs.map(_.asUInt).reduce(_|_)),
+    alloc_way
+  )
+  override val metaSz = s2_meta.asUInt.getWidth
+
+  // update
+  val s1_update_cfi_idx = s1_update.bits.cfi_idx.bits
+  val s1_update_meta    = s1_update.bits.meta.asTypeOf(new IBTBPredictMeta)
+
+  val s1_update_wibtb_data = Wire(new IBTBEntry)
+  s1_update_wibtb_data.target := s1_update.bits.target.asUInt
+  val s1_update_wibtb_mask = (
+    UIntToOH(s1_update_cfi_idx) &
+    Fill(bankWidth,
+      s1_update.valid &&
+      s1_update.bits.cfi_idx.valid &&
+      s1_update.bits.cfi_is_jalr &&
+      s1_update.bits.cfi_mispredicted &&
+      s1_update.bits.is_commit_update
+    )
+  )
+  val s1_update_wmeta_mask = s1_update_wibtb_mask
+  val s1_update_wmeta_data = Wire(Vec(bankWidth, new IBTBMeta))
+  for (w <- 0 until bankWidth) {
+    s1_update_wmeta_data(w).tag := s1_update_meta.write_tag
+  }
+
+  for (w <- 0 until nWays) {
+    when (doing_reset || s1_update_meta.write_way === w.U || (w == 0 && nWays == 1).B) {
+      // write variables
+      val ibtb_widx = Mux(doing_reset, reset_idx, s1_update_meta.write_tag(log2Ceil(nSets)-1,0))
+      val meta_widx = Mux(doing_reset, reset_idx, s1_update_meta.write_tag(log2Ceil(nSets)-1,0))
+      val ibtb_mask = Mux(doing_reset, ~(0.U(bankWidth.W)), s1_update_wibtb_mask).asBools
+      val meta_mask = Mux(doing_reset, ~(0.U(bankWidth.W)), s1_update_wmeta_mask).asBools
+      val ibtb_wdat = Mux(
+        doing_reset,
+        VecInit(Seq.fill(bankWidth) { 0.U(ibtbEntrySz.W) }),
+        VecInit(Seq.fill(bankWidth) { s1_update_wibtb_data.asUInt })
+      )
+      val meta_wdat = Mux(
+        doing_reset,
+        VecInit(Seq.fill(bankWidth) { 0.U(ibtbMetaSz.W) }),
+        VecInit(s1_update_wmeta_data.map(_.asUInt))
+      )
+      // write entry
+      ibtb(w).write(ibtb_widx, ibtb_wdat, ibtb_mask)
+      meta(w).write(meta_widx, meta_wdat, meta_mask)
+    }
+  }
+}
diff --git a/src/main/scala/ifu/bpd/predictor.scala b/src/main/scala/ifu/bpd/predictor.scala
index f176c8df..2ad7448e 100644
--- a/src/main/scala/ifu/bpd/predictor.scala
+++ b/src/main/scala/ifu/bpd/predictor.scala
@@ -144,6 +144,7 @@ abstract class BranchPredictorBank(implicit p: Parameters) extends BoomModule()(
     // Local history not available until end of f1
     val f1_ghist = Input(UInt(globalHistoryLength.W))
     val f1_lhist = Input(UInt(localHistoryLength.W))
+    val f1_phist = Input(UInt((indirectHistoryLength * indirectHistoryWidth).W))
 
     val resp_in = Input(Vec(nInputs, new BranchPredictionBankResponse))
     val resp = Output(new BranchPredictionBankResponse)
@@ -238,6 +239,7 @@ class BranchPredictor(implicit p: Parameters) extends BoomModule()(p)
 
     banked_predictors(0).io.f1_ghist := RegNext(io.f0_req.bits.ghist.histories(0))
     banked_predictors(0).io.f1_lhist := banked_lhist_providers(0).io.f1_lhist
+    banked_predictors(0).io.f1_phist := RegNext(io.f0_req.bits.ghist.path_history)
 
     banked_predictors(0).io.resp_in(0)           := (0.U).asTypeOf(new BranchPredictionBankResponse)
   } else {
@@ -249,6 +251,9 @@ class BranchPredictor(implicit p: Parameters) extends BoomModule()(p)
     banked_predictors(0).io.f1_lhist := banked_lhist_providers(0).io.f1_lhist
     banked_predictors(1).io.f1_lhist := banked_lhist_providers(1).io.f1_lhist
 
+    banked_predictors(0).io.f1_phist  := RegNext(io.f0_req.bits.ghist.path_history)
+    banked_predictors(1).io.f1_phist  := RegNext(io.f0_req.bits.ghist.path_history)
+
     when (bank(io.f0_req.bits.pc) === 0.U) {
       banked_lhist_providers(0).io.f0_valid := io.f0_req.valid
       banked_lhist_providers(0).io.f0_pc    := bankAlign(io.f0_req.bits.pc)
diff --git a/src/main/scala/ifu/fetch-target-queue.scala b/src/main/scala/ifu/fetch-target-queue.scala
index 5af74617..886a8281 100644
--- a/src/main/scala/ifu/fetch-target-queue.scala
+++ b/src/main/scala/ifu/fetch-target-queue.scala
@@ -294,6 +294,11 @@ class FetchTargetQueue(implicit p: Parameters) extends BoomModule
     io.bpdupdate.bits.target     := bpd_target
     io.bpdupdate.bits.cfi_is_br  := bpd_entry.br_mask(cfi_idx)
     io.bpdupdate.bits.cfi_is_jal := bpd_entry.cfi_type === CFI_JAL || bpd_entry.cfi_type === CFI_JALR
+    io.bpdupdate.bits.cfi_is_jalr := bpd_entry.cfi_type === CFI_JALR && !bpd_entry.cfi_is_ret &&
+    // The unequality condition is redundant because it is true when do_commit_update/is_commit_update
+    //   is true, but it results in different behavior if not added. Even just asserting the condition
+    //   can avoid this issue, possibly a Chisel bug.
+      RegNext(WrapInc(bpd_idx, num_entries) =/= enq_ptr)
     io.bpdupdate.bits.ghist      := bpd_ghist
     io.bpdupdate.bits.lhist      := bpd_lhist
     io.bpdupdate.bits.meta       := bpd_meta
diff --git a/src/main/scala/ifu/frontend.scala b/src/main/scala/ifu/frontend.scala
index 17bf3b2e..a8281bc4 100644
--- a/src/main/scala/ifu/frontend.scala
+++ b/src/main/scala/ifu/frontend.scala
@@ -57,6 +57,8 @@ class GlobalHistory(implicit p: Parameters) extends BoomBundle()(p)
 
   val ras_idx = UInt(log2Ceil(nRasEntries).W)
 
+  val path_history = UInt((indirectHistoryLength * indirectHistoryWidth).W)
+
   def histories(bank: Int) = {
     if (nBanks == 1) {
       old_history
@@ -82,7 +84,8 @@ class GlobalHistory(implicit p: Parameters) extends BoomBundle()(p)
 
   def update(branches: UInt, cfi_taken: Bool, cfi_is_br: Bool, cfi_idx: UInt,
     cfi_valid: Bool, addr: UInt,
-    cfi_is_call: Bool, cfi_is_ret: Bool): GlobalHistory = {
+    cfi_is_call: Bool, cfi_is_ret: Bool,
+    cfi_is_jalr: Bool = false.B, cfi_target: UInt = 0.U): GlobalHistory = {
     val cfi_idx_fixed = cfi_idx(log2Ceil(fetchWidth)-1,0)
     val cfi_idx_oh = UIntToOH(cfi_idx_fixed)
     val new_history = Wire(new GlobalHistory)
@@ -123,6 +126,10 @@ class GlobalHistory(implicit p: Parameters) extends BoomBundle()(p)
     }
     new_history.ras_idx := Mux(cfi_valid && cfi_is_call, WrapInc(ras_idx, nRasEntries),
                            Mux(cfi_valid && cfi_is_ret , WrapDec(ras_idx, nRasEntries), ras_idx))
+    new_history.path_history := Mux(
+      cfi_valid && cfi_is_jalr,
+      path_history << indirectHistoryWidth | (cfi_target & (1.U << indirectHistoryWidth) - 1.U),
+      path_history)
     new_history
   }
 
@@ -286,6 +293,11 @@ class BoomFrontendIO(implicit p: Parameters) extends BoomBundle
   val flush_icache = Output(Bool())
 
   val perf = Input(new FrontendPerfEvents)
+
+  val tperf = Input(new Bundle {
+    val tlb_miss = Vec(ftqSz, Bool())
+    val cache_miss = Vec(ftqSz, Bool())
+  })
 }
 
 /**
@@ -799,7 +811,9 @@ class BoomFrontendModule(outer: BoomFrontend) extends LazyModuleImp(outer)
     f3_fetch_bundle.cfi_idx.valid,
     f3_fetch_bundle.pc,
     f3_fetch_bundle.cfi_is_call,
-    f3_fetch_bundle.cfi_is_ret
+    f3_fetch_bundle.cfi_is_ret,
+    f3_fetch_bundle.cfi_type === CFI_JALR && !f3_fetch_bundle.cfi_is_ret,
+    f3_fetch_bundle.next_pc
   )
 
 
@@ -985,6 +999,46 @@ class BoomFrontendModule(outer: BoomFrontend) extends LazyModuleImp(outer)
   ftq.io.debug_ftq_idx := io.cpu.debug_ftq_idx
   io.cpu.debug_fetch_pc := ftq.io.debug_fetch_pc
 
+  // jxy: performance monitor
+  val tlbVadd = RegInit(0.U(vaddrBitsExtended.W))
+  val cachePadd = RegInit(0.U(paddrBits.W))
+  val f3_tlbMiss = withReset(reset.asBool || f3_clear) { Module(new Queue(Bool(), 1, pipe=true, flow=false)) }
+  val f3_cacheMiss = withReset(reset.asBool || f3_clear) { Module(new Queue(Bool(), 1, pipe=true, flow=false)) }
+  val f4_tlbMiss = withReset(reset.asBool || f4_clear) { Module(new Queue(Bool(), 1, pipe=true, flow=false)) }
+  val f4_cacheMiss = withReset(reset.asBool || f4_clear) { Module(new Queue(Bool(), 1, pipe=true, flow=false)) }
+  val ftq_enq_ptr = RegInit(1.U(log2Ceil(ftqSz).W))
+  val ftq_tlbMiss = Reg(Vec(ftqSz, Bool()))
+  val ftq_cacheMiss = Reg(Vec(ftqSz, Bool()))
+  f3_tlbMiss  .io.enq.valid := f3.io.enq.valid
+  f3_cacheMiss.io.enq.valid := f3.io.enq.valid
+  f3_tlbMiss  .io.enq.bits  := (s2_vpc >> pgIdxBits)    ===  tlbVadd
+  f3_cacheMiss.io.enq.bits  := (s2_ppc >> blockOffBits) === (cachePadd >> blockOffBits)
+  f3_tlbMiss  .io.deq.ready := f3.io.deq.ready
+  f3_cacheMiss.io.deq.ready := f3.io.deq.ready
+  f4_tlbMiss  .io.enq.valid := f4.io.enq.valid
+  f4_cacheMiss.io.enq.valid := f4.io.enq.valid
+  f4_tlbMiss  .io.enq.bits  := f3_tlbMiss  .io.deq.bits
+  f4_cacheMiss.io.enq.bits  := f3_cacheMiss.io.deq.bits
+  f4_tlbMiss  .io.deq.ready := f4.io.deq.ready
+  f4_cacheMiss.io.deq.ready := f4.io.deq.ready
+  when (f3_tlbMiss  .io.enq.valid && f3_tlbMiss  .io.enq.bits) { tlbVadd   := 0.U }
+  when (f3_cacheMiss.io.enq.valid && f3_cacheMiss.io.enq.bits) { cachePadd := 0.U }
+  when (tlb   .io.perf.miss)    { tlbVadd   := tlb   .io.perf.vadd }
+  when (icache.io.perf.acquire) { cachePadd := icache.io.perf.padd }
+  when (ftq.io.deq.valid) {
+    ftq_tlbMiss  (ftq.io.deq.bits) := false.B
+    ftq_cacheMiss(ftq.io.deq.bits) := false.B
+  }
+  when (ftq.io.enq.fire) {
+    ftq_tlbMiss  (ftq_enq_ptr) := f4_tlbMiss  .io.deq.bits
+    ftq_cacheMiss(ftq_enq_ptr) := f4_cacheMiss.io.deq.bits
+    ftq_enq_ptr := WrapInc(ftq_enq_ptr, ftqSz)
+  }
+  when (ftq.io.redirect.valid) {
+    ftq_enq_ptr := WrapInc(ftq.io.redirect.bits, ftqSz)
+  }
+  io.cpu.tperf.tlb_miss   := ftq_tlbMiss
+  io.cpu.tperf.cache_miss := ftq_cacheMiss
 
   override def toString: String =
     (BoomCoreStringPrefix("====Overall Frontend Params====") + "\n"
diff --git a/src/main/scala/ifu/icache.scala b/src/main/scala/ifu/icache.scala
index 916a688f..e7181fc1 100644
--- a/src/main/scala/ifu/icache.scala
+++ b/src/main/scala/ifu/icache.scala
@@ -82,6 +82,7 @@ class ICacheBundle(val outer: ICache) extends BoomBundle()(outer.p)
 
   val perf = Output(new Bundle {
     val acquire = Bool()
+    val padd = UInt(paddrBits.W)
   })
 }
 
@@ -334,6 +335,7 @@ class ICacheModule(outer: ICache) extends LazyModuleImp(outer)
   tl_out.e.valid := false.B
 
   io.perf.acquire := tl_out.a.fire
+  io.perf.padd := refill_paddr
 
   when (!refill_valid) { invalidated := false.B }
   when (refill_fire) { refill_valid := true.B }
diff --git a/src/main/scala/lsu/dcache.scala b/src/main/scala/lsu/dcache.scala
index 9f326f50..64fae1b0 100644
--- a/src/main/scala/lsu/dcache.scala
+++ b/src/main/scala/lsu/dcache.scala
@@ -820,6 +820,7 @@ class BoomNonBlockingDCacheModule(outer: BoomNonBlockingDCache) extends LazyModu
 
   io.lsu.perf.release := edge.done(tl_out.c)
   io.lsu.perf.acquire := edge.done(tl_out.a)
+  io.lsu.perf.padd := tl_out.a.bits.address
 
   // load data gen
   val s2_data_word_prebypass = widthMap(w => s2_data_muxed(w) >> Cat(s2_word_idx(w), 0.U(log2Ceil(coreDataBits).W)))
diff --git a/src/main/scala/lsu/lsu.scala b/src/main/scala/lsu/lsu.scala
index 1eb8ba24..ef57c55d 100644
--- a/src/main/scala/lsu/lsu.scala
+++ b/src/main/scala/lsu/lsu.scala
@@ -102,6 +102,7 @@ class LSUDMemIO(implicit p: Parameters, edge: TLEdgeOut) extends BoomBundle()(p)
   val perf = Input(new Bundle {
     val acquire = Bool()
     val release = Bool()
+    val padd = UInt(paddrBits.W)
   })
 
 }
@@ -153,6 +154,11 @@ class LSUCoreIO(implicit p: Parameters) extends BoomBundle()(p)
     val release = Bool()
     val tlbMiss = Bool()
   })
+
+  val tperf = Output(Vec(memWidth, new Bundle {
+    val tlb_miss = Valid(UInt(robAddrSz.W))
+    val cache_miss = Valid(UInt(robAddrSz.W))
+  }))
 }
 
 class LSUIO(implicit p: Parameters, edge: TLEdgeOut) extends BoomBundle()(p)
@@ -1651,6 +1657,32 @@ class LSU(implicit p: Parameters, edge: TLEdgeOut) extends BoomModule()(p)
                     ~(st_exc_killed_mask.asUInt)
 
 
+  // jxy: performance monitor
+  val tlbVadd = RegInit((0.U(vaddrBitsExtended.W)))
+  val cachePadd = RegInit(0.U(coreMaxAddrBits.W))
+  for (w <- 0 until memWidth) {
+    io.core.tperf(w).tlb_miss.valid := false.B
+    io.core.tperf(w).cache_miss.valid := false.B
+    io.core.tperf(w).tlb_miss.bits := DontCare
+    io.core.tperf(w).cache_miss.bits := DontCare
+    when (exe_tlb_valid(w) && (exe_tlb_vaddr(w) >> pgIdxBits) === tlbVadd) {
+      io.core.tperf(w).tlb_miss.valid := true.B
+      io.core.tperf(w).tlb_miss.bits := exe_tlb_uop(w).rob_idx
+      tlbVadd := 0.U
+    }
+    val resp_addr = Mux(
+      io.dmem.resp(w).bits.uop.uses_ldq,
+      ldq(io.dmem.resp(w).bits.uop.ldq_idx).bits.addr.bits,
+      stq(io.dmem.resp(w).bits.uop.stq_idx).bits.addr.bits
+    )
+    when (io.dmem.resp(w).valid && (resp_addr >> blockOffBits) === (cachePadd >> blockOffBits)) {
+      io.core.tperf(w).cache_miss.valid := true.B
+      io.core.tperf(w).cache_miss.bits := io.dmem.resp(w).bits.uop.rob_idx
+      cachePadd := 0.U
+    }
+  }
+  when (dtlb.io.perf.miss)    { tlbVadd   := dtlb.io.perf.vadd }
+  when (io.dmem.perf.acquire) { cachePadd := io.dmem.perf.padd }
 }
 
 /**
diff --git a/src/main/scala/lsu/tlb.scala b/src/main/scala/lsu/tlb.scala
index a09018b4..311b0412 100644
--- a/src/main/scala/lsu/tlb.scala
+++ b/src/main/scala/lsu/tlb.scala
@@ -23,6 +23,11 @@ class NBDTLB(instruction: Boolean, lgMaxSize: Int, cfg: TLBConfig)(implicit edge
     val sfence = Input(Valid(new SFenceReq))
     val ptw = new TLBPTWIO
     val kill = Input(Bool())
+    // jxy: performance monitor
+    val perf = Output(new Bundle {
+      val miss = Output(Bool())
+      val vadd = Output(UInt(vaddrBitsExtended.W))
+    })
   })
   io.ptw := DontCare
   io.resp := DontCare
@@ -358,4 +363,7 @@ class NBDTLB(instruction: Boolean, lgMaxSize: Int, cfg: TLBConfig)(implicit edge
   }
 
 
+  // jxy: performance monitor
+  io.perf.miss := io.ptw.req.fire
+  io.perf.vadd := io.ptw.req.bits.bits.addr
 }
