diff --git a/arch/riscv/kernel/asm-offsets.c b/arch/riscv/kernel/asm-offsets.c
index 16490755304e..5b939b3ce115 100644
--- a/arch/riscv/kernel/asm-offsets.c
+++ b/arch/riscv/kernel/asm-offsets.c
@@ -20,6 +20,9 @@ void asm_offsets(void);
 
 void asm_offsets(void)
 {
+	OFFSET(TASK_PID, task_struct, pid);
+	OFFSET(TASK_TGID, task_struct, tgid);
+
 	OFFSET(TASK_THREAD_RA, task_struct, thread.ra);
 	OFFSET(TASK_THREAD_SP, task_struct, thread.sp);
 	OFFSET(TASK_THREAD_S0, task_struct, thread.s[0]);
diff --git a/arch/riscv/kernel/entry.S b/arch/riscv/kernel/entry.S
index 33a5a9f2a0d4..7a8d3c4f6d04 100644
--- a/arch/riscv/kernel/entry.S
+++ b/arch/riscv/kernel/entry.S
@@ -399,6 +399,12 @@ SYM_FUNC_START(__switch_to)
 	REG_S s11, TASK_THREAD_S11_RA(a3)
 	/* Save the kernel shadow call stack pointer */
 	scs_save_current
+	/* jxy: HINT instruction to send PID info */
+	lw s1, TASK_TGID(a1) // a1: next task_struct
+	lw s2, TASK_PID(a1)
+	slli s2, s2, 32
+	or s1, s1, s2
+	slli zero, s1, 0
 	/* Restore context from next->thread */
 	REG_L ra,  TASK_THREAD_RA_RA(a4)
 	REG_L sp,  TASK_THREAD_SP_RA(a4)
