diff --git a/board/genesys2/bootrom.dts b/board/genesys2/bootrom.dts
index 1a41eee..9b5ae0f 100644
--- a/board/genesys2/bootrom.dts
+++ b/board/genesys2/bootrom.dts
@@ -8,6 +8,29 @@
         stdout-path = "serial0";
     };
 
+    pmu {
+        compatible = "riscv,pmu";
+        riscv,event-to-mhpmevent =
+            <0x00004 0x00000000 0x0302>,
+            <0x10001 0x00000000 0x0202>,
+            <0x10002 0x00000000 0x0402>,
+            <0x10009 0x00000000 0x0102>,
+            <0x10011 0x00000000 0x2002>,
+            <0x10019 0x00000000 0x1002>,
+            <0x10021 0x00000000 0x0802>;
+        riscv,event-to-mhpmcounters =
+            <0x00004 0x00004 0x18>,
+            <0x10001 0x10002 0x18>,
+            <0x10009 0x10009 0x18>,
+            <0x10011 0x10011 0x18>,
+            <0x10019 0x10019 0x18>,
+            <0x10021 0x10021 0x18>;
+        riscv,raw-event-to-mhpmcounters =
+            <0x0 0x0 0xffffffff 0xfc0000ff 0x18>,
+            <0x0 0x1 0xffffffff 0xffffc0ff 0x18>,
+            <0x0 0x2 0xffffffff 0xffffc0ff 0x18>;
+    };
+
     io-bus {
         #address-cells = <1>;
         #size-cells = <1>;
diff --git a/patches/opensbi/Kconfig b/patches/opensbi/Kconfig
index aa62595..5418142 100644
--- a/patches/opensbi/Kconfig
+++ b/patches/opensbi/Kconfig
@@ -3,6 +3,8 @@
 config PLATFORM_VIVADO_RISC_V
 	bool
 	select FDT
+	select FDT_DOMAIN
+	select FDT_PMU
 	select IPI_MSWI
 	select IRQCHIP_PLIC
 	select TIMER_MTIMER
diff --git a/patches/opensbi/platform.c b/patches/opensbi/platform.c
index b69a3fa..30283ec 100644
--- a/patches/opensbi/platform.c
+++ b/patches/opensbi/platform.c
@@ -20,6 +20,7 @@
 #include <sbi_utils/cppc/fdt_cppc.h>
 #include <sbi_utils/hsm/fdt_hsm.h>
 #include <sbi_utils/ipi/fdt_ipi.h>
+#include <sbi_utils/fdt/fdt_pmu.h>
 
 #define SR_RX_FIFO_VALID_DATA   (1 << 0) /* data in receive FIFO */
 #define SR_RX_FIFO_FULL         (1 << 1) /* receive FIFO full    */
@@ -111,11 +112,47 @@ static int final_init(bool cold_boot) {
     return 0;
 }
 
+static int pmu_init(void)
+{
+    int rc;
+
+    rc = fdt_pmu_setup(fdt_get_address());
+    if (rc && rc != SBI_ENOENT)
+        return rc;
+
+    return 0;
+}
+
+static uint64_t pmu_xlate_to_mhpmevent(uint32_t event_idx,
+					       uint64_t data)
+{
+	uint64_t evt_val = 0;
+
+	/* data is valid only for raw events and is equal to event selector */
+	if (event_idx == SBI_PMU_EVENT_RAW_IDX ||
+		event_idx == SBI_PMU_EVENT_RAW_V2_IDX)
+		evt_val = data;
+	else {
+		/**
+		 * Generic platform follows the SBI specification recommendation
+		 * i.e. zero extended event_idx is used as mhpmevent value for
+		 * hardware general/cache events if platform does't define one.
+		 */
+		evt_val = fdt_pmu_get_select_value(event_idx);
+		if (!evt_val)
+			evt_val = (uint64_t)event_idx;
+	}
+
+	return evt_val;
+}
+
 const struct sbi_platform_operations platform_ops = {
     .early_init = early_init,
     .final_init = final_init,
     .irqchip_init = fdt_irqchip_init,
     .ipi_init = fdt_ipi_init,
+	.pmu_init = pmu_init,
+	.pmu_xlate_to_mhpmevent = pmu_xlate_to_mhpmevent,
     .timer_init = fdt_timer_init,
 };
 
